// rclnodejs/generated/teleop_tools_msgs/teleop_tools_msgs__action__Increment.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class teleop_tools_msgs__action__Increment {
  static get Goal() {
    return require('./teleop_tools_msgs__action__Increment_Goal.js');
  }
  static get Result() {
    return require('./teleop_tools_msgs__action__Increment_Result.js');
  }
  static get Feedback() {
    return require('./teleop_tools_msgs__action__Increment_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./teleop_tools_msgs__action__Increment_SendGoal.js');
      },
      get GetResultService() {
        return require('./teleop_tools_msgs__action__Increment_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./teleop_tools_msgs__action__Increment_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'teleop_tools_msgs', subFolder: 'action', interfaceName: 'Increment'};
  }
}
module.exports = teleop_tools_msgs__action__Increment;
