// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__DetectMoreLoopClosures_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DetectMoreLoopClosures_ResponseRefStruct = StructType({
  detected: primitiveTypes.int32,
});
const DetectMoreLoopClosures_ResponseRefArray = ArrayType(DetectMoreLoopClosures_ResponseRefStruct);
const DetectMoreLoopClosures_ResponseRefStructArray = StructType({
  data: DetectMoreLoopClosures_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DetectMoreLoopClosures_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DetectMoreLoopClosures_ResponseRefStruct(other._refObject.toObject());
      this._detectedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DetectMoreLoopClosures_ResponseRefStruct();
    this._detectedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DetectMoreLoopClosures_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DetectMoreLoopClosures_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return DetectMoreLoopClosures_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return DetectMoreLoopClosures_ResponseRefStructArray
  }
  static get refObjectType() {
    return DetectMoreLoopClosures_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._detectedIntialized) {
        throw new TypeError('Invalid argument: detected in DetectMoreLoopClosures_Response');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._detectedIntialized = true;
    this._refObject.detected = refObject.detected;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DetectMoreLoopClosures_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'DetectMoreLoopClosures_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get detected() {
    if (!this._detectedIntialized) {
      return undefined;
    }
    return this._refObject.detected;
  }
  set detected(value) {
    this._detectedIntialized = true;
    this._refObject.detected = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DetectMoreLoopClosures_ResponseRefStruct(refObject.toObject());
    this._detectedIntialized = true;
  }
  copy(other) {
    this._refObject = new DetectMoreLoopClosures_ResponseRefStruct(other._refObject.toObject());
    this._detectedIntialized = true;
  }
  static get classType() {
    return DetectMoreLoopClosures_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"detected","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"DetectMoreLoopClosures_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DetectMoreLoopClosures_Response"};
  }
  hasMember(name) {
    let memberNames = ["detected"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DetectMoreLoopClosures_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DetectMoreLoopClosures_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DetectMoreLoopClosures_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DetectMoreLoopClosures_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DetectMoreLoopClosures_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DetectMoreLoopClosures_ResponseRefArray(size);
    this._refObject = new DetectMoreLoopClosures_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DetectMoreLoopClosures_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DetectMoreLoopClosures_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DetectMoreLoopClosures_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DetectMoreLoopClosures_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DetectMoreLoopClosures_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DetectMoreLoopClosures_ResponseArrayWrapper;
  }
}
module.exports = DetectMoreLoopClosures_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "detected",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "DetectMoreLoopClosures_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DetectMoreLoopClosures_Response"
}
*/
