// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GetNodesInRadius_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const GetNodesInRadius_RequestRefStruct = StructType({
  node_id: primitiveTypes.int32,
  x: primitiveTypes.float32,
  y: primitiveTypes.float32,
  z: primitiveTypes.float32,
  radius: primitiveTypes.float32,
  k: primitiveTypes.int32,
});
const GetNodesInRadius_RequestRefArray = ArrayType(GetNodesInRadius_RequestRefStruct);
const GetNodesInRadius_RequestRefStructArray = StructType({
  data: GetNodesInRadius_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetNodesInRadius_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetNodesInRadius_RequestRefStruct(other._refObject.toObject());
      this._node_idIntialized = true;
      this._xIntialized = true;
      this._yIntialized = true;
      this._zIntialized = true;
      this._radiusIntialized = true;
      this._kIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetNodesInRadius_RequestRefStruct();
    this._node_idIntialized = false;
    this._xIntialized = false;
    this._yIntialized = false;
    this._zIntialized = false;
    this._radiusIntialized = false;
    this._kIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new GetNodesInRadius_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetNodesInRadius_RequestArrayWrapper;
  }
  static get ArrayType() {
    return GetNodesInRadius_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetNodesInRadius_RequestRefStructArray
  }
  static get refObjectType() {
    return GetNodesInRadius_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._node_idIntialized) {
        throw new TypeError('Invalid argument: node_id in GetNodesInRadius_Request');
      }
      if (checkConsistency && !this._xIntialized) {
        throw new TypeError('Invalid argument: x in GetNodesInRadius_Request');
      }
      if (checkConsistency && !this._yIntialized) {
        throw new TypeError('Invalid argument: y in GetNodesInRadius_Request');
      }
      if (checkConsistency && !this._zIntialized) {
        throw new TypeError('Invalid argument: z in GetNodesInRadius_Request');
      }
      if (checkConsistency && !this._radiusIntialized) {
        throw new TypeError('Invalid argument: radius in GetNodesInRadius_Request');
      }
      if (checkConsistency && !this._kIntialized) {
        throw new TypeError('Invalid argument: k in GetNodesInRadius_Request');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._node_idIntialized = true;
    this._refObject.node_id = refObject.node_id;
    this._xIntialized = true;
    this._refObject.x = refObject.x;
    this._yIntialized = true;
    this._refObject.y = refObject.y;
    this._zIntialized = true;
    this._refObject.z = refObject.z;
    this._radiusIntialized = true;
    this._refObject.radius = refObject.radius;
    this._kIntialized = true;
    this._refObject.k = refObject.k;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    GetNodesInRadius_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GetNodesInRadius_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_id() {
    if (!this._node_idIntialized) {
      return undefined;
    }
    return this._refObject.node_id;
  }
  set node_id(value) {
    this._node_idIntialized = true;
    this._refObject.node_id = value;
  }
  get x() {
    if (!this._xIntialized) {
      return undefined;
    }
    return this._refObject.x;
  }
  set x(value) {
    this._xIntialized = true;
    this._refObject.x = value;
  }
  get y() {
    if (!this._yIntialized) {
      return undefined;
    }
    return this._refObject.y;
  }
  set y(value) {
    this._yIntialized = true;
    this._refObject.y = value;
  }
  get z() {
    if (!this._zIntialized) {
      return undefined;
    }
    return this._refObject.z;
  }
  set z(value) {
    this._zIntialized = true;
    this._refObject.z = value;
  }
  get radius() {
    if (!this._radiusIntialized) {
      return undefined;
    }
    return this._refObject.radius;
  }
  set radius(value) {
    this._radiusIntialized = true;
    this._refObject.radius = value;
  }
  get k() {
    if (!this._kIntialized) {
      return undefined;
    }
    return this._refObject.k;
  }
  set k(value) {
    this._kIntialized = true;
    this._refObject.k = value;
  }
  copyRefObject(refObject) {
    this._refObject = new GetNodesInRadius_RequestRefStruct(refObject.toObject());
    this._node_idIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._radiusIntialized = true;
    this._kIntialized = true;
  }
  copy(other) {
    this._refObject = new GetNodesInRadius_RequestRefStruct(other._refObject.toObject());
    this._node_idIntialized = true;
    this._xIntialized = true;
    this._yIntialized = true;
    this._zIntialized = true;
    this._radiusIntialized = true;
    this._kIntialized = true;
  }
  static get classType() {
    return GetNodesInRadius_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"x","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"y","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"z","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"k","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GetNodesInRadius_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetNodesInRadius_Request"};
  }
  hasMember(name) {
    let memberNames = ["node_id","x","y","z","radius","k"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetNodesInRadius_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetNodesInRadius_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetNodesInRadius_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetNodesInRadius_RequestRefArray(size);
    this._refObject = new GetNodesInRadius_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetNodesInRadius_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetNodesInRadius_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetNodesInRadius_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetNodesInRadius_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetNodesInRadius_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetNodesInRadius_RequestArrayWrapper;
  }
}
module.exports = GetNodesInRadius_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "x",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "y",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "z",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "k",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "GetNodesInRadius_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetNodesInRadius_Request"
}
*/
