// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__RemoveLabel_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const RemoveLabel_RequestRefStruct = StructType({
  label: primitiveTypes.string,
});
const RemoveLabel_RequestRefArray = ArrayType(RemoveLabel_RequestRefStruct);
const RemoveLabel_RequestRefStructArray = StructType({
  data: RemoveLabel_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RemoveLabel_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RemoveLabel_RequestRefStruct(other._refObject.toObject());
      this._labelIntialized = true;
      this._wrapperFields.label =  new StringWrapper(other._wrapperFields.label);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RemoveLabel_RequestRefStruct();
    this._labelIntialized = false;
    this._wrapperFields.label = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new RemoveLabel_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RemoveLabel_RequestArrayWrapper;
  }
  static get ArrayType() {
    return RemoveLabel_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return RemoveLabel_RequestRefStructArray
  }
  static get refObjectType() {
    return RemoveLabel_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._labelIntialized) {
        throw new TypeError('Invalid argument: label in RemoveLabel_Request');
      }
    if (own) {
      this._wrapperFields.label.freeze(own, checkConsistency);
    }
    this._refObject.label = this._wrapperFields.label.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._labelIntialized = true;
    this._wrapperFields.label.data = refObject.label.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.label);
  }
  static destoryRawROS(msg) {
    RemoveLabel_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'RemoveLabel_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get label() {
    if (!this._labelIntialized) {
      return undefined;
    }
    return this._wrapperFields.label.data;
  }
  set label(value) {
    this._labelIntialized = true;
    this._wrapperFields.label.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RemoveLabel_RequestRefStruct(refObject.toObject());
    this._labelIntialized = true;
    this._wrapperFields.label.copyRefObject(this._refObject.label);
  }
  copy(other) {
    this._refObject = new RemoveLabel_RequestRefStruct(other._refObject.toObject());
    this._labelIntialized = true;
    this._wrapperFields.label.copy(other._wrapperFields.label);
  }
  static get classType() {
    return RemoveLabel_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"RemoveLabel_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RemoveLabel_Request"};
  }
  hasMember(name) {
    let memberNames = ["label"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RemoveLabel_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RemoveLabel_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RemoveLabel_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RemoveLabel_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RemoveLabel_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RemoveLabel_RequestRefArray(size);
    this._refObject = new RemoveLabel_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RemoveLabel_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RemoveLabel_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RemoveLabel_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RemoveLabel_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RemoveLabel_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RemoveLabel_RequestArrayWrapper;
  }
}
module.exports = RemoveLabel_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "label",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "RemoveLabel_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RemoveLabel_Request"
}
*/
