// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__PublishMap_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PublishMap_RequestRefStruct = StructType({
  global_map: primitiveTypes.bool,
  optimized: primitiveTypes.bool,
  graph_only: primitiveTypes.bool,
});
const PublishMap_RequestRefArray = ArrayType(PublishMap_RequestRefStruct);
const PublishMap_RequestRefStructArray = StructType({
  data: PublishMap_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class PublishMap_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new PublishMap_RequestRefStruct(other._refObject.toObject());
      this._global_mapIntialized = true;
      this._optimizedIntialized = true;
      this._graph_onlyIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new PublishMap_RequestRefStruct();
    this._global_mapIntialized = false;
    this._optimizedIntialized = false;
    this._graph_onlyIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new PublishMap_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new PublishMap_RequestArrayWrapper;
  }
  static get ArrayType() {
    return PublishMap_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return PublishMap_RequestRefStructArray
  }
  static get refObjectType() {
    return PublishMap_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._global_mapIntialized) {
        throw new TypeError('Invalid argument: global_map in PublishMap_Request');
      }
      if (checkConsistency && !this._optimizedIntialized) {
        throw new TypeError('Invalid argument: optimized in PublishMap_Request');
      }
      if (checkConsistency && !this._graph_onlyIntialized) {
        throw new TypeError('Invalid argument: graph_only in PublishMap_Request');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._global_mapIntialized = true;
    this._refObject.global_map = refObject.global_map;
    this._optimizedIntialized = true;
    this._refObject.optimized = refObject.optimized;
    this._graph_onlyIntialized = true;
    this._refObject.graph_only = refObject.graph_only;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    PublishMap_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'PublishMap_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get global_map() {
    if (!this._global_mapIntialized) {
      return undefined;
    }
    return this._refObject.global_map;
  }
  set global_map(value) {
    this._global_mapIntialized = true;
    this._refObject.global_map = value;
  }
  get optimized() {
    if (!this._optimizedIntialized) {
      return undefined;
    }
    return this._refObject.optimized;
  }
  set optimized(value) {
    this._optimizedIntialized = true;
    this._refObject.optimized = value;
  }
  get graph_only() {
    if (!this._graph_onlyIntialized) {
      return undefined;
    }
    return this._refObject.graph_only;
  }
  set graph_only(value) {
    this._graph_onlyIntialized = true;
    this._refObject.graph_only = value;
  }
  copyRefObject(refObject) {
    this._refObject = new PublishMap_RequestRefStruct(refObject.toObject());
    this._global_mapIntialized = true;
    this._optimizedIntialized = true;
    this._graph_onlyIntialized = true;
  }
  copy(other) {
    this._refObject = new PublishMap_RequestRefStruct(other._refObject.toObject());
    this._global_mapIntialized = true;
    this._optimizedIntialized = true;
    this._graph_onlyIntialized = true;
  }
  static get classType() {
    return PublishMap_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"global_map","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"optimized","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"graph_only","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"PublishMap_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"PublishMap_Request"};
  }
  hasMember(name) {
    let memberNames = ["global_map","optimized","graph_only"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class PublishMap_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof PublishMap_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new PublishMap_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PublishMap_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to PublishMap_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new PublishMap_RequestRefArray(size);
    this._refObject = new PublishMap_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new PublishMap_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof PublishMap_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "PublishMap_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      PublishMap_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return PublishMap_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return PublishMap_RequestArrayWrapper;
  }
}
module.exports = PublishMap_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "global_map",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "optimized",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "graph_only",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "PublishMap_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "PublishMap_Request"
}
*/
