// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__SetGoal_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const SetGoal_RequestRefStruct = StructType({
  node_id: primitiveTypes.int32,
  node_label: primitiveTypes.string,
  frame_id: primitiveTypes.string,
});
const SetGoal_RequestRefArray = ArrayType(SetGoal_RequestRefStruct);
const SetGoal_RequestRefStructArray = StructType({
  data: SetGoal_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetGoal_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetGoal_RequestRefStruct(other._refObject.toObject());
      this._node_idIntialized = true;
      this._node_labelIntialized = true;
      this._wrapperFields.node_label =  new StringWrapper(other._wrapperFields.node_label);
      this._frame_idIntialized = true;
      this._wrapperFields.frame_id =  new StringWrapper(other._wrapperFields.frame_id);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetGoal_RequestRefStruct();
    this._node_idIntialized = false;
    this._node_labelIntialized = false;
    this._wrapperFields.node_label = new StringWrapper();
    this._frame_idIntialized = false;
    this._wrapperFields.frame_id = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new SetGoal_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetGoal_RequestArrayWrapper;
  }
  static get ArrayType() {
    return SetGoal_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetGoal_RequestRefStructArray
  }
  static get refObjectType() {
    return SetGoal_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._node_idIntialized) {
        throw new TypeError('Invalid argument: node_id in SetGoal_Request');
      }
      if (checkConsistency && !this._node_labelIntialized) {
        throw new TypeError('Invalid argument: node_label in SetGoal_Request');
      }
      if (checkConsistency && !this._frame_idIntialized) {
        throw new TypeError('Invalid argument: frame_id in SetGoal_Request');
      }
    if (own) {
      this._wrapperFields.node_label.freeze(own, checkConsistency);
    }
    this._refObject.node_label = this._wrapperFields.node_label.refObject;
    if (own) {
      this._wrapperFields.frame_id.freeze(own, checkConsistency);
    }
    this._refObject.frame_id = this._wrapperFields.frame_id.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._node_idIntialized = true;
    this._refObject.node_id = refObject.node_id;
    this._node_labelIntialized = true;
    this._wrapperFields.node_label.data = refObject.node_label.data;
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = refObject.frame_id.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.node_label);
    StringWrapper.freeStruct(refObject.frame_id);
  }
  static destoryRawROS(msg) {
    SetGoal_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'SetGoal_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get node_id() {
    if (!this._node_idIntialized) {
      return undefined;
    }
    return this._refObject.node_id;
  }
  set node_id(value) {
    this._node_idIntialized = true;
    this._refObject.node_id = value;
  }
  get node_label() {
    if (!this._node_labelIntialized) {
      return undefined;
    }
    return this._wrapperFields.node_label.data;
  }
  set node_label(value) {
    this._node_labelIntialized = true;
    this._wrapperFields.node_label.data = value;
  }
  get frame_id() {
    if (!this._frame_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.frame_id.data;
  }
  set frame_id(value) {
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetGoal_RequestRefStruct(refObject.toObject());
    this._node_idIntialized = true;
    this._node_labelIntialized = true;
    this._wrapperFields.node_label.copyRefObject(this._refObject.node_label);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copyRefObject(this._refObject.frame_id);
  }
  copy(other) {
    this._refObject = new SetGoal_RequestRefStruct(other._refObject.toObject());
    this._node_idIntialized = true;
    this._node_labelIntialized = true;
    this._wrapperFields.node_label.copy(other._wrapperFields.node_label);
    this._frame_idIntialized = true;
    this._wrapperFields.frame_id.copy(other._wrapperFields.frame_id);
  }
  static get classType() {
    return SetGoal_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"node_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"node_label","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"frame_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"SetGoal_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetGoal_Request"};
  }
  hasMember(name) {
    let memberNames = ["node_id","node_label","frame_id"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetGoal_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetGoal_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetGoal_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGoal_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGoal_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetGoal_RequestRefArray(size);
    this._refObject = new SetGoal_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetGoal_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetGoal_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetGoal_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetGoal_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetGoal_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetGoal_RequestArrayWrapper;
  }
}
module.exports = SetGoal_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "node_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "node_label",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "frame_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "SetGoal_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetGoal_Request"
}
*/
