// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__LoadDatabase_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const LoadDatabase_RequestRefStruct = StructType({
  database_path: primitiveTypes.string,
  clear: primitiveTypes.bool,
});
const LoadDatabase_RequestRefArray = ArrayType(LoadDatabase_RequestRefStruct);
const LoadDatabase_RequestRefStructArray = StructType({
  data: LoadDatabase_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LoadDatabase_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LoadDatabase_RequestRefStruct(other._refObject.toObject());
      this._database_pathIntialized = true;
      this._wrapperFields.database_path =  new StringWrapper(other._wrapperFields.database_path);
      this._clearIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LoadDatabase_RequestRefStruct();
    this._database_pathIntialized = false;
    this._wrapperFields.database_path = new StringWrapper();
    this._clearIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LoadDatabase_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LoadDatabase_RequestArrayWrapper;
  }
  static get ArrayType() {
    return LoadDatabase_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return LoadDatabase_RequestRefStructArray
  }
  static get refObjectType() {
    return LoadDatabase_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._database_pathIntialized) {
        throw new TypeError('Invalid argument: database_path in LoadDatabase_Request');
      }
      if (checkConsistency && !this._clearIntialized) {
        throw new TypeError('Invalid argument: clear in LoadDatabase_Request');
      }
    if (own) {
      this._wrapperFields.database_path.freeze(own, checkConsistency);
    }
    this._refObject.database_path = this._wrapperFields.database_path.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._database_pathIntialized = true;
    this._wrapperFields.database_path.data = refObject.database_path.data;
    this._clearIntialized = true;
    this._refObject.clear = refObject.clear;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.database_path);
  }
  static destoryRawROS(msg) {
    LoadDatabase_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'LoadDatabase_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get database_path() {
    if (!this._database_pathIntialized) {
      return undefined;
    }
    return this._wrapperFields.database_path.data;
  }
  set database_path(value) {
    this._database_pathIntialized = true;
    this._wrapperFields.database_path.data = value;
  }
  get clear() {
    if (!this._clearIntialized) {
      return undefined;
    }
    return this._refObject.clear;
  }
  set clear(value) {
    this._clearIntialized = true;
    this._refObject.clear = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LoadDatabase_RequestRefStruct(refObject.toObject());
    this._database_pathIntialized = true;
    this._wrapperFields.database_path.copyRefObject(this._refObject.database_path);
    this._clearIntialized = true;
  }
  copy(other) {
    this._refObject = new LoadDatabase_RequestRefStruct(other._refObject.toObject());
    this._database_pathIntialized = true;
    this._wrapperFields.database_path.copy(other._wrapperFields.database_path);
    this._clearIntialized = true;
  }
  static get classType() {
    return LoadDatabase_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"database_path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"clear","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"LoadDatabase_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LoadDatabase_Request"};
  }
  hasMember(name) {
    let memberNames = ["database_path","clear"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LoadDatabase_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LoadDatabase_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LoadDatabase_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadDatabase_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LoadDatabase_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LoadDatabase_RequestRefArray(size);
    this._refObject = new LoadDatabase_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LoadDatabase_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LoadDatabase_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LoadDatabase_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LoadDatabase_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LoadDatabase_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LoadDatabase_RequestArrayWrapper;
  }
}
module.exports = LoadDatabase_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "database_path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "clear",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "LoadDatabase_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LoadDatabase_Request"
}
*/
