// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__IntrusionDatum.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const IntrusionDatumRefStruct = StructType({
  size: primitiveTypes.uint32,
  flags: BoolWrapper.refObjectArrayType,
});
const IntrusionDatumRefArray = ArrayType(IntrusionDatumRefStruct);
const IntrusionDatumRefStructArray = StructType({
  data: IntrusionDatumRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IntrusionDatumWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._flagsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IntrusionDatumRefStruct(other._refObject.toObject());
      this._sizeIntialized = true;
      this._wrapperFields.flags = BoolWrapper.createArray();
      this._wrapperFields.flags.copy(other._wrapperFields.flags);
      this.flags = other.flags;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IntrusionDatumRefStruct();
    this._sizeIntialized = false;
    this._wrapperFields.flags = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new IntrusionDatumWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IntrusionDatumArrayWrapper;
  }
  static get ArrayType() {
    return IntrusionDatumArrayWrapper;
  }
  static get refObjectArrayType() {
    return IntrusionDatumRefStructArray
  }
  static get refObjectType() {
    return IntrusionDatumRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._sizeIntialized) {
        throw new TypeError('Invalid argument: size in IntrusionDatum');
      }
    if (!own) {
      this._wrapperFields.flags.fill(this._flagsArray);
      this._wrapperFields.flags.freeze(own, checkConsistency);
      this._refObject.flags = this._wrapperFields.flags.refObject;
    } else {
      this._wrapperFields.flags.fill([]);
      this._wrapperFields.flags.freeze(own, checkConsistency);
      this._refObject.flags = this._wrapperFields.flags.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._sizeIntialized = true;
    this._refObject.size = refObject.size;
    refObject.flags.data.length = refObject.flags.size;
    for (let index = 0; index < refObject.flags.size; index++) {
      this._flagsArray[index] = refObject.flags.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.flags.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.flags);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.flags, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    IntrusionDatumWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'IntrusionDatum'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get size() {
    if (!this._sizeIntialized) {
      return undefined;
    }
    return this._refObject.size;
  }
  set size(value) {
    this._sizeIntialized = true;
    this._refObject.size = value;
  }
  get flags() {
    return this._flagsArray;
  }
  set flags(value) {
    this._flagsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new IntrusionDatumRefStruct(refObject.toObject());
    this._sizeIntialized = true;
    refObject.flags.data.length = refObject.flags.size;
    for (let index = 0; index < refObject.flags.size; index++) {
      this._flagsArray[index] = refObject.flags.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new IntrusionDatumRefStruct(other._refObject.toObject());
    this._sizeIntialized = true;
    this._flagsArray = other._flagsArray.slice();
  }
  static get classType() {
    return IntrusionDatumWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"size","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"flags","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"IntrusionDatum","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IntrusionDatum"};
  }
  hasMember(name) {
    let memberNames = ["size","flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IntrusionDatumArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IntrusionDatumWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IntrusionDatumWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntrusionDatumArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IntrusionDatumArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IntrusionDatumRefArray(size);
    this._refObject = new IntrusionDatumRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IntrusionDatumWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IntrusionDatumArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IntrusionDatumArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IntrusionDatumWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IntrusionDatumWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IntrusionDatumArrayWrapper;
  }
}
module.exports = IntrusionDatumWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "size",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "flags",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "IntrusionDatum",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IntrusionDatum"
}
*/
