// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__GetNodesInRadius_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const GetNodesInRadius_ResponseRefStruct = StructType({
  ids: Int32Wrapper.refObjectArrayType,
  poses: PoseWrapper.refObjectArrayType,
  dists_sqr: Float32Wrapper.refObjectArrayType,
});
const GetNodesInRadius_ResponseRefArray = ArrayType(GetNodesInRadius_ResponseRefStruct);
const GetNodesInRadius_ResponseRefStructArray = StructType({
  data: GetNodesInRadius_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetNodesInRadius_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetNodesInRadius_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.ids = Int32Wrapper.createArray();
      this._wrapperFields.ids.copy(other._wrapperFields.ids);
      this._wrapperFields.poses = PoseWrapper.createArray();
      this._wrapperFields.poses.copy(other._wrapperFields.poses);
      this._wrapperFields.dists_sqr = Float32Wrapper.createArray();
      this._wrapperFields.dists_sqr.copy(other._wrapperFields.dists_sqr);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetNodesInRadius_ResponseRefStruct();
    this._wrapperFields.ids = Int32Wrapper.createArray();
    this._wrapperFields.poses = PoseWrapper.createArray();
    this._wrapperFields.dists_sqr = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new GetNodesInRadius_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetNodesInRadius_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetNodesInRadius_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetNodesInRadius_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetNodesInRadius_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.ids.freeze(own, checkConsistency);
    this._refObject.ids = this._wrapperFields.ids.refObject;
    if (own) {
      this._wrapperFields.ids.fill(Int32Array.from([]));
      this._wrapperFields.ids.freeze(own, checkConsistency);
      this._refObject.ids = this._wrapperFields.ids.refObject;
    }
    this._wrapperFields.poses.freeze(own, checkConsistency);
    this._refObject.poses = this._wrapperFields.poses.refObject;
    this._wrapperFields.dists_sqr.freeze(own, checkConsistency);
    this._refObject.dists_sqr = this._wrapperFields.dists_sqr.refObject;
    if (own) {
      this._wrapperFields.dists_sqr.fill(Float32Array.from([]));
      this._wrapperFields.dists_sqr.freeze(own, checkConsistency);
      this._refObject.dists_sqr = this._wrapperFields.dists_sqr.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ids.copyRefObject(refObject.ids);
    this._wrapperFields.poses.copyRefObject(refObject.poses);
    this._wrapperFields.dists_sqr.copyRefObject(refObject.dists_sqr);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ids.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.ids);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ids, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.poses.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.poses);
      if (PoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.poses, PoseWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.dists_sqr.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.dists_sqr);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.dists_sqr, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    GetNodesInRadius_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'GetNodesInRadius_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ids() {
    return this._wrapperFields['ids'].data;
  }
  set ids(value) {
    this._wrapperFields['ids'].fill(value);
  }
  get poses() {
    return this._wrapperFields.poses;
  }
  set poses(value) {
    this._wrapperFields.poses.fill(value);
  }
  get dists_sqr() {
    return this._wrapperFields['dists_sqr'].data;
  }
  set dists_sqr(value) {
    this._wrapperFields['dists_sqr'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new GetNodesInRadius_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.ids.copyRefObject(this._refObject.ids);
    this._wrapperFields.poses.copyRefObject(this._refObject.poses);
    this._wrapperFields.dists_sqr.copyRefObject(this._refObject.dists_sqr);
  }
  copy(other) {
    this._refObject = new GetNodesInRadius_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.ids.copy(other._wrapperFields.ids);
    this._wrapperFields.poses.copy(other._wrapperFields.poses);
    this._wrapperFields.dists_sqr.copy(other._wrapperFields.dists_sqr);
  }
  static get classType() {
    return GetNodesInRadius_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"dists_sqr","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"GetNodesInRadius_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetNodesInRadius_Response"};
  }
  hasMember(name) {
    let memberNames = ["ids","poses","dists_sqr"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetNodesInRadius_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetNodesInRadius_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetNodesInRadius_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetNodesInRadius_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetNodesInRadius_ResponseRefArray(size);
    this._refObject = new GetNodesInRadius_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetNodesInRadius_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetNodesInRadius_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetNodesInRadius_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetNodesInRadius_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetNodesInRadius_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetNodesInRadius_ResponseArrayWrapper;
  }
}
module.exports = GetNodesInRadius_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "dists_sqr",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "GetNodesInRadius_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetNodesInRadius_Response"
}
*/
