// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__MonitoringCase.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const MonitoringCaseRefStruct = StructType({
  monitoring_case_number: primitiveTypes.int32,
  fields: Int32Wrapper.refObjectArrayType,
  fields_valid: BoolWrapper.refObjectArrayType,
});
const MonitoringCaseRefArray = ArrayType(MonitoringCaseRefStruct);
const MonitoringCaseRefStructArray = StructType({
  data: MonitoringCaseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MonitoringCaseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._fields_validArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MonitoringCaseRefStruct(other._refObject.toObject());
      this._monitoring_case_numberIntialized = true;
      this._wrapperFields.fields = Int32Wrapper.createArray();
      this._wrapperFields.fields.copy(other._wrapperFields.fields);
      this._wrapperFields.fields_valid = BoolWrapper.createArray();
      this._wrapperFields.fields_valid.copy(other._wrapperFields.fields_valid);
      this.fields_valid = other.fields_valid;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MonitoringCaseRefStruct();
    this._monitoring_case_numberIntialized = false;
    this._wrapperFields.fields = Int32Wrapper.createArray();
    this._wrapperFields.fields_valid = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MonitoringCaseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MonitoringCaseArrayWrapper;
  }
  static get ArrayType() {
    return MonitoringCaseArrayWrapper;
  }
  static get refObjectArrayType() {
    return MonitoringCaseRefStructArray
  }
  static get refObjectType() {
    return MonitoringCaseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._monitoring_case_numberIntialized) {
        throw new TypeError('Invalid argument: monitoring_case_number in MonitoringCase');
      }
    this._wrapperFields.fields.freeze(own, checkConsistency);
    this._refObject.fields = this._wrapperFields.fields.refObject;
    if (own) {
      this._wrapperFields.fields.fill(Int32Array.from([]));
      this._wrapperFields.fields.freeze(own, checkConsistency);
      this._refObject.fields = this._wrapperFields.fields.refObject;
    }
    if (!own) {
      this._wrapperFields.fields_valid.fill(this._fields_validArray);
      this._wrapperFields.fields_valid.freeze(own, checkConsistency);
      this._refObject.fields_valid = this._wrapperFields.fields_valid.refObject;
    } else {
      this._wrapperFields.fields_valid.fill([]);
      this._wrapperFields.fields_valid.freeze(own, checkConsistency);
      this._refObject.fields_valid = this._wrapperFields.fields_valid.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._monitoring_case_numberIntialized = true;
    this._refObject.monitoring_case_number = refObject.monitoring_case_number;
    this._wrapperFields.fields.copyRefObject(refObject.fields);
    refObject.fields_valid.data.length = refObject.fields_valid.size;
    for (let index = 0; index < refObject.fields_valid.size; index++) {
      this._fields_validArray[index] = refObject.fields_valid.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.fields.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.fields);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fields, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.fields_valid.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.fields_valid);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.fields_valid, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MonitoringCaseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'MonitoringCase'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get monitoring_case_number() {
    if (!this._monitoring_case_numberIntialized) {
      return undefined;
    }
    return this._refObject.monitoring_case_number;
  }
  set monitoring_case_number(value) {
    this._monitoring_case_numberIntialized = true;
    this._refObject.monitoring_case_number = value;
  }
  get fields() {
    return this._wrapperFields['fields'].data;
  }
  set fields(value) {
    this._wrapperFields['fields'].fill(value);
  }
  get fields_valid() {
    return this._fields_validArray;
  }
  set fields_valid(value) {
    this._fields_validArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MonitoringCaseRefStruct(refObject.toObject());
    this._monitoring_case_numberIntialized = true;
    this._wrapperFields.fields.copyRefObject(this._refObject.fields);
    refObject.fields_valid.data.length = refObject.fields_valid.size;
    for (let index = 0; index < refObject.fields_valid.size; index++) {
      this._fields_validArray[index] = refObject.fields_valid.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new MonitoringCaseRefStruct(other._refObject.toObject());
    this._monitoring_case_numberIntialized = true;
    this._wrapperFields.fields.copy(other._wrapperFields.fields);
    this._fields_validArray = other._fields_validArray.slice();
  }
  static get classType() {
    return MonitoringCaseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"monitoring_case_number","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fields","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"fields_valid","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"MonitoringCase","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MonitoringCase"};
  }
  hasMember(name) {
    let memberNames = ["monitoring_case_number","fields","fields_valid"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MonitoringCaseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MonitoringCaseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MonitoringCaseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MonitoringCaseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MonitoringCaseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MonitoringCaseRefArray(size);
    this._refObject = new MonitoringCaseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MonitoringCaseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MonitoringCaseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MonitoringCaseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MonitoringCaseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MonitoringCaseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MonitoringCaseArrayWrapper;
  }
}
module.exports = MonitoringCaseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "monitoring_case_number",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fields",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "fields_valid",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "MonitoringCase",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MonitoringCase"
}
*/
