// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ApplicationData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ApplicationInputsWrapper = require('../../generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ApplicationInputs.js');
const ApplicationOutputsWrapper = require('../../generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ApplicationOutputs.js');
const ApplicationDataRefStruct = StructType({
  inputs: ApplicationInputsWrapper.refObjectType,
  outputs: ApplicationOutputsWrapper.refObjectType,
});
const ApplicationDataRefArray = ArrayType(ApplicationDataRefStruct);
const ApplicationDataRefStructArray = StructType({
  data: ApplicationDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ApplicationDataWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ApplicationDataRefStruct(other._refObject.toObject());
      this._wrapperFields.inputs =  new ApplicationInputsWrapper(other._wrapperFields.inputs);
      this._wrapperFields.outputs =  new ApplicationOutputsWrapper(other._wrapperFields.outputs);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ApplicationDataRefStruct();
    this._wrapperFields.inputs = new ApplicationInputsWrapper();
    this._wrapperFields.outputs = new ApplicationOutputsWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new ApplicationDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ApplicationDataArrayWrapper;
  }
  static get ArrayType() {
    return ApplicationDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return ApplicationDataRefStructArray
  }
  static get refObjectType() {
    return ApplicationDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.inputs.freeze(own, checkConsistency);
    this._refObject.inputs = this._wrapperFields.inputs.refObject;
    this._wrapperFields.outputs.freeze(own, checkConsistency);
    this._refObject.outputs = this._wrapperFields.outputs.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.inputs.copyRefObject(refObject.inputs);
    this._wrapperFields.outputs.copyRefObject(refObject.outputs);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    ApplicationInputsWrapper.freeStruct(refObject.inputs);
    ApplicationOutputsWrapper.freeStruct(refObject.outputs);
  }
  static destoryRawROS(msg) {
    ApplicationDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'ApplicationData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get inputs() {
    return this._wrapperFields.inputs;
  }
  set inputs(value) {
    if (value instanceof ApplicationInputsWrapper) {
      this._wrapperFields.inputs.copy(value);
    } else {
      this._wrapperFields.inputs.copy(new ApplicationInputsWrapper(value));
    }
  }
  get outputs() {
    return this._wrapperFields.outputs;
  }
  set outputs(value) {
    if (value instanceof ApplicationOutputsWrapper) {
      this._wrapperFields.outputs.copy(value);
    } else {
      this._wrapperFields.outputs.copy(new ApplicationOutputsWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new ApplicationDataRefStruct(refObject.toObject());
    this._wrapperFields.inputs.copyRefObject(this._refObject.inputs);
    this._wrapperFields.outputs.copyRefObject(this._refObject.outputs);
  }
  copy(other) {
    this._refObject = new ApplicationDataRefStruct(other._refObject.toObject());
    this._wrapperFields.inputs.copy(other._wrapperFields.inputs);
    this._wrapperFields.outputs.copy(other._wrapperFields.outputs);
  }
  static get classType() {
    return ApplicationDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"inputs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sick_safetyscanners2_interfaces","type":"ApplicationInputs","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"outputs","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"sick_safetyscanners2_interfaces","type":"ApplicationOutputs","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"ApplicationData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ApplicationData"};
  }
  hasMember(name) {
    let memberNames = ["inputs","outputs"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ApplicationDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ApplicationDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ApplicationDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ApplicationDataRefArray(size);
    this._refObject = new ApplicationDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ApplicationDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ApplicationDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ApplicationDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ApplicationDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ApplicationDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ApplicationDataArrayWrapper;
  }
}
module.exports = ApplicationDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "inputs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sick_safetyscanners2_interfaces",
        "type": "ApplicationInputs",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "outputs",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "sick_safetyscanners2_interfaces",
        "type": "ApplicationOutputs",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "ApplicationData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ApplicationData"
}
*/
