// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ApplicationOutputs.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const Int16Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int16.js');
const ApplicationOutputsRefStruct = StructType({
  evaluation_path_outputs_eval_out: BoolWrapper.refObjectArrayType,
  evaluation_path_outputs_is_safe: BoolWrapper.refObjectArrayType,
  evaluation_path_outputs_is_valid: BoolWrapper.refObjectArrayType,
  monitoring_case_number_outputs: UInt16Wrapper.refObjectArrayType,
  monitoring_case_number_outputs_flags: BoolWrapper.refObjectArrayType,
  sleep_mode_output: primitiveTypes.uint8,
  sleep_mode_output_valid: primitiveTypes.bool,
  error_flag_contamination_warning: primitiveTypes.bool,
  error_flag_contamination_error: primitiveTypes.bool,
  error_flag_manipulation_error: primitiveTypes.bool,
  error_flag_glare: primitiveTypes.bool,
  error_flag_reference_contour_intruded: primitiveTypes.bool,
  error_flag_critical_error: primitiveTypes.bool,
  error_flags_are_valid: primitiveTypes.bool,
  linear_velocity_outputs_velocity_0: primitiveTypes.int16,
  linear_velocity_outputs_velocity_0_valid: primitiveTypes.bool,
  linear_velocity_outputs_velocity_0_transmitted_safely: primitiveTypes.bool,
  linear_velocity_outputs_velocity_1: primitiveTypes.int16,
  linear_velocity_outputs_velocity_1_valid: primitiveTypes.bool,
  linear_velocity_outputs_velocity_1_transmitted_safely: primitiveTypes.bool,
  resulting_velocity: Int16Wrapper.refObjectArrayType,
  resulting_velocity_flags: BoolWrapper.refObjectArrayType,
});
const ApplicationOutputsRefArray = ArrayType(ApplicationOutputsRefStruct);
const ApplicationOutputsRefStructArray = StructType({
  data: ApplicationOutputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ApplicationOutputsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._evaluation_path_outputs_eval_outArray = [];
    this._evaluation_path_outputs_is_safeArray = [];
    this._evaluation_path_outputs_is_validArray = [];
    this._monitoring_case_number_outputs_flagsArray = [];
    this._resulting_velocity_flagsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ApplicationOutputsRefStruct(other._refObject.toObject());
      this._wrapperFields.evaluation_path_outputs_eval_out = BoolWrapper.createArray();
      this._wrapperFields.evaluation_path_outputs_eval_out.copy(other._wrapperFields.evaluation_path_outputs_eval_out);
      this.evaluation_path_outputs_eval_out = other.evaluation_path_outputs_eval_out;
      this._wrapperFields.evaluation_path_outputs_is_safe = BoolWrapper.createArray();
      this._wrapperFields.evaluation_path_outputs_is_safe.copy(other._wrapperFields.evaluation_path_outputs_is_safe);
      this.evaluation_path_outputs_is_safe = other.evaluation_path_outputs_is_safe;
      this._wrapperFields.evaluation_path_outputs_is_valid = BoolWrapper.createArray();
      this._wrapperFields.evaluation_path_outputs_is_valid.copy(other._wrapperFields.evaluation_path_outputs_is_valid);
      this.evaluation_path_outputs_is_valid = other.evaluation_path_outputs_is_valid;
      this._wrapperFields.monitoring_case_number_outputs = UInt16Wrapper.createArray();
      this._wrapperFields.monitoring_case_number_outputs.copy(other._wrapperFields.monitoring_case_number_outputs);
      this._wrapperFields.monitoring_case_number_outputs_flags = BoolWrapper.createArray();
      this._wrapperFields.monitoring_case_number_outputs_flags.copy(other._wrapperFields.monitoring_case_number_outputs_flags);
      this.monitoring_case_number_outputs_flags = other.monitoring_case_number_outputs_flags;
      this._sleep_mode_outputIntialized = true;
      this._sleep_mode_output_validIntialized = true;
      this._error_flag_contamination_warningIntialized = true;
      this._error_flag_contamination_errorIntialized = true;
      this._error_flag_manipulation_errorIntialized = true;
      this._error_flag_glareIntialized = true;
      this._error_flag_reference_contour_intrudedIntialized = true;
      this._error_flag_critical_errorIntialized = true;
      this._error_flags_are_validIntialized = true;
      this._linear_velocity_outputs_velocity_0Intialized = true;
      this._linear_velocity_outputs_velocity_0_validIntialized = true;
      this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = true;
      this._linear_velocity_outputs_velocity_1Intialized = true;
      this._linear_velocity_outputs_velocity_1_validIntialized = true;
      this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = true;
      this._wrapperFields.resulting_velocity = Int16Wrapper.createArray();
      this._wrapperFields.resulting_velocity.copy(other._wrapperFields.resulting_velocity);
      this._wrapperFields.resulting_velocity_flags = BoolWrapper.createArray();
      this._wrapperFields.resulting_velocity_flags.copy(other._wrapperFields.resulting_velocity_flags);
      this.resulting_velocity_flags = other.resulting_velocity_flags;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ApplicationOutputsRefStruct();
    this._wrapperFields.evaluation_path_outputs_eval_out = BoolWrapper.createArray();
    this._wrapperFields.evaluation_path_outputs_is_safe = BoolWrapper.createArray();
    this._wrapperFields.evaluation_path_outputs_is_valid = BoolWrapper.createArray();
    this._wrapperFields.monitoring_case_number_outputs = UInt16Wrapper.createArray();
    this._wrapperFields.monitoring_case_number_outputs_flags = BoolWrapper.createArray();
    this._sleep_mode_outputIntialized = false;
    this._sleep_mode_output_validIntialized = false;
    this._error_flag_contamination_warningIntialized = false;
    this._error_flag_contamination_errorIntialized = false;
    this._error_flag_manipulation_errorIntialized = false;
    this._error_flag_glareIntialized = false;
    this._error_flag_reference_contour_intrudedIntialized = false;
    this._error_flag_critical_errorIntialized = false;
    this._error_flags_are_validIntialized = false;
    this._linear_velocity_outputs_velocity_0Intialized = false;
    this._linear_velocity_outputs_velocity_0_validIntialized = false;
    this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = false;
    this._linear_velocity_outputs_velocity_1Intialized = false;
    this._linear_velocity_outputs_velocity_1_validIntialized = false;
    this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = false;
    this._wrapperFields.resulting_velocity = Int16Wrapper.createArray();
    this._wrapperFields.resulting_velocity_flags = BoolWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ApplicationOutputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ApplicationOutputsArrayWrapper;
  }
  static get ArrayType() {
    return ApplicationOutputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ApplicationOutputsRefStructArray
  }
  static get refObjectType() {
    return ApplicationOutputsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._sleep_mode_outputIntialized) {
        throw new TypeError('Invalid argument: sleep_mode_output in ApplicationOutputs');
      }
      if (checkConsistency && !this._sleep_mode_output_validIntialized) {
        throw new TypeError('Invalid argument: sleep_mode_output_valid in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_contamination_warningIntialized) {
        throw new TypeError('Invalid argument: error_flag_contamination_warning in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_contamination_errorIntialized) {
        throw new TypeError('Invalid argument: error_flag_contamination_error in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_manipulation_errorIntialized) {
        throw new TypeError('Invalid argument: error_flag_manipulation_error in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_glareIntialized) {
        throw new TypeError('Invalid argument: error_flag_glare in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_reference_contour_intrudedIntialized) {
        throw new TypeError('Invalid argument: error_flag_reference_contour_intruded in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flag_critical_errorIntialized) {
        throw new TypeError('Invalid argument: error_flag_critical_error in ApplicationOutputs');
      }
      if (checkConsistency && !this._error_flags_are_validIntialized) {
        throw new TypeError('Invalid argument: error_flags_are_valid in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_0Intialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_0 in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_0_validIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_0_valid in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_0_transmitted_safely in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_1Intialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_1 in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_1_validIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_1_valid in ApplicationOutputs');
      }
      if (checkConsistency && !this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_outputs_velocity_1_transmitted_safely in ApplicationOutputs');
      }
    if (!own) {
      this._wrapperFields.evaluation_path_outputs_eval_out.fill(this._evaluation_path_outputs_eval_outArray);
      this._wrapperFields.evaluation_path_outputs_eval_out.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_eval_out = this._wrapperFields.evaluation_path_outputs_eval_out.refObject;
    } else {
      this._wrapperFields.evaluation_path_outputs_eval_out.fill([]);
      this._wrapperFields.evaluation_path_outputs_eval_out.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_eval_out = this._wrapperFields.evaluation_path_outputs_eval_out.refObject;
    }
    if (!own) {
      this._wrapperFields.evaluation_path_outputs_is_safe.fill(this._evaluation_path_outputs_is_safeArray);
      this._wrapperFields.evaluation_path_outputs_is_safe.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_is_safe = this._wrapperFields.evaluation_path_outputs_is_safe.refObject;
    } else {
      this._wrapperFields.evaluation_path_outputs_is_safe.fill([]);
      this._wrapperFields.evaluation_path_outputs_is_safe.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_is_safe = this._wrapperFields.evaluation_path_outputs_is_safe.refObject;
    }
    if (!own) {
      this._wrapperFields.evaluation_path_outputs_is_valid.fill(this._evaluation_path_outputs_is_validArray);
      this._wrapperFields.evaluation_path_outputs_is_valid.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_is_valid = this._wrapperFields.evaluation_path_outputs_is_valid.refObject;
    } else {
      this._wrapperFields.evaluation_path_outputs_is_valid.fill([]);
      this._wrapperFields.evaluation_path_outputs_is_valid.freeze(own, checkConsistency);
      this._refObject.evaluation_path_outputs_is_valid = this._wrapperFields.evaluation_path_outputs_is_valid.refObject;
    }
    this._wrapperFields.monitoring_case_number_outputs.freeze(own, checkConsistency);
    this._refObject.monitoring_case_number_outputs = this._wrapperFields.monitoring_case_number_outputs.refObject;
    if (own) {
      this._wrapperFields.monitoring_case_number_outputs.fill(Uint16Array.from([]));
      this._wrapperFields.monitoring_case_number_outputs.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_outputs = this._wrapperFields.monitoring_case_number_outputs.refObject;
    }
    if (!own) {
      this._wrapperFields.monitoring_case_number_outputs_flags.fill(this._monitoring_case_number_outputs_flagsArray);
      this._wrapperFields.monitoring_case_number_outputs_flags.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_outputs_flags = this._wrapperFields.monitoring_case_number_outputs_flags.refObject;
    } else {
      this._wrapperFields.monitoring_case_number_outputs_flags.fill([]);
      this._wrapperFields.monitoring_case_number_outputs_flags.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_outputs_flags = this._wrapperFields.monitoring_case_number_outputs_flags.refObject;
    }
    this._wrapperFields.resulting_velocity.freeze(own, checkConsistency);
    this._refObject.resulting_velocity = this._wrapperFields.resulting_velocity.refObject;
    if (own) {
      this._wrapperFields.resulting_velocity.fill(Int16Array.from([]));
      this._wrapperFields.resulting_velocity.freeze(own, checkConsistency);
      this._refObject.resulting_velocity = this._wrapperFields.resulting_velocity.refObject;
    }
    if (!own) {
      this._wrapperFields.resulting_velocity_flags.fill(this._resulting_velocity_flagsArray);
      this._wrapperFields.resulting_velocity_flags.freeze(own, checkConsistency);
      this._refObject.resulting_velocity_flags = this._wrapperFields.resulting_velocity_flags.refObject;
    } else {
      this._wrapperFields.resulting_velocity_flags.fill([]);
      this._wrapperFields.resulting_velocity_flags.freeze(own, checkConsistency);
      this._refObject.resulting_velocity_flags = this._wrapperFields.resulting_velocity_flags.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.evaluation_path_outputs_eval_out.data.length = refObject.evaluation_path_outputs_eval_out.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_eval_out.size; index++) {
      this._evaluation_path_outputs_eval_outArray[index] = refObject.evaluation_path_outputs_eval_out.data[index].data;
    }
    refObject.evaluation_path_outputs_is_safe.data.length = refObject.evaluation_path_outputs_is_safe.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_is_safe.size; index++) {
      this._evaluation_path_outputs_is_safeArray[index] = refObject.evaluation_path_outputs_is_safe.data[index].data;
    }
    refObject.evaluation_path_outputs_is_valid.data.length = refObject.evaluation_path_outputs_is_valid.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_is_valid.size; index++) {
      this._evaluation_path_outputs_is_validArray[index] = refObject.evaluation_path_outputs_is_valid.data[index].data;
    }
    this._wrapperFields.monitoring_case_number_outputs.copyRefObject(refObject.monitoring_case_number_outputs);
    refObject.monitoring_case_number_outputs_flags.data.length = refObject.monitoring_case_number_outputs_flags.size;
    for (let index = 0; index < refObject.monitoring_case_number_outputs_flags.size; index++) {
      this._monitoring_case_number_outputs_flagsArray[index] = refObject.monitoring_case_number_outputs_flags.data[index].data;
    }
    this._sleep_mode_outputIntialized = true;
    this._refObject.sleep_mode_output = refObject.sleep_mode_output;
    this._sleep_mode_output_validIntialized = true;
    this._refObject.sleep_mode_output_valid = refObject.sleep_mode_output_valid;
    this._error_flag_contamination_warningIntialized = true;
    this._refObject.error_flag_contamination_warning = refObject.error_flag_contamination_warning;
    this._error_flag_contamination_errorIntialized = true;
    this._refObject.error_flag_contamination_error = refObject.error_flag_contamination_error;
    this._error_flag_manipulation_errorIntialized = true;
    this._refObject.error_flag_manipulation_error = refObject.error_flag_manipulation_error;
    this._error_flag_glareIntialized = true;
    this._refObject.error_flag_glare = refObject.error_flag_glare;
    this._error_flag_reference_contour_intrudedIntialized = true;
    this._refObject.error_flag_reference_contour_intruded = refObject.error_flag_reference_contour_intruded;
    this._error_flag_critical_errorIntialized = true;
    this._refObject.error_flag_critical_error = refObject.error_flag_critical_error;
    this._error_flags_are_validIntialized = true;
    this._refObject.error_flags_are_valid = refObject.error_flags_are_valid;
    this._linear_velocity_outputs_velocity_0Intialized = true;
    this._refObject.linear_velocity_outputs_velocity_0 = refObject.linear_velocity_outputs_velocity_0;
    this._linear_velocity_outputs_velocity_0_validIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_0_valid = refObject.linear_velocity_outputs_velocity_0_valid;
    this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_0_transmitted_safely = refObject.linear_velocity_outputs_velocity_0_transmitted_safely;
    this._linear_velocity_outputs_velocity_1Intialized = true;
    this._refObject.linear_velocity_outputs_velocity_1 = refObject.linear_velocity_outputs_velocity_1;
    this._linear_velocity_outputs_velocity_1_validIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_1_valid = refObject.linear_velocity_outputs_velocity_1_valid;
    this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_1_transmitted_safely = refObject.linear_velocity_outputs_velocity_1_transmitted_safely;
    this._wrapperFields.resulting_velocity.copyRefObject(refObject.resulting_velocity);
    refObject.resulting_velocity_flags.data.length = refObject.resulting_velocity_flags.size;
    for (let index = 0; index < refObject.resulting_velocity_flags.size; index++) {
      this._resulting_velocity_flagsArray[index] = refObject.resulting_velocity_flags.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.evaluation_path_outputs_eval_out.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.evaluation_path_outputs_eval_out);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.evaluation_path_outputs_eval_out, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.evaluation_path_outputs_is_safe.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.evaluation_path_outputs_is_safe);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.evaluation_path_outputs_is_safe, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.evaluation_path_outputs_is_valid.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.evaluation_path_outputs_is_valid);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.evaluation_path_outputs_is_valid, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.monitoring_case_number_outputs.size != 0) {
      UInt16Wrapper.ArrayType.freeArray(refObject.monitoring_case_number_outputs);
      if (UInt16Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.monitoring_case_number_outputs, UInt16Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.monitoring_case_number_outputs_flags.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.monitoring_case_number_outputs_flags);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.monitoring_case_number_outputs_flags, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.resulting_velocity.size != 0) {
      Int16Wrapper.ArrayType.freeArray(refObject.resulting_velocity);
      if (Int16Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.resulting_velocity, Int16Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.resulting_velocity_flags.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.resulting_velocity_flags);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.resulting_velocity_flags, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ApplicationOutputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'ApplicationOutputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get evaluation_path_outputs_eval_out() {
    return this._evaluation_path_outputs_eval_outArray;
  }
  set evaluation_path_outputs_eval_out(value) {
    this._evaluation_path_outputs_eval_outArray = value;
  }
  get evaluation_path_outputs_is_safe() {
    return this._evaluation_path_outputs_is_safeArray;
  }
  set evaluation_path_outputs_is_safe(value) {
    this._evaluation_path_outputs_is_safeArray = value;
  }
  get evaluation_path_outputs_is_valid() {
    return this._evaluation_path_outputs_is_validArray;
  }
  set evaluation_path_outputs_is_valid(value) {
    this._evaluation_path_outputs_is_validArray = value;
  }
  get monitoring_case_number_outputs() {
    return this._wrapperFields['monitoring_case_number_outputs'].data;
  }
  set monitoring_case_number_outputs(value) {
    this._wrapperFields['monitoring_case_number_outputs'].fill(value);
  }
  get monitoring_case_number_outputs_flags() {
    return this._monitoring_case_number_outputs_flagsArray;
  }
  set monitoring_case_number_outputs_flags(value) {
    this._monitoring_case_number_outputs_flagsArray = value;
  }
  get sleep_mode_output() {
    if (!this._sleep_mode_outputIntialized) {
      return undefined;
    }
    return this._refObject.sleep_mode_output;
  }
  set sleep_mode_output(value) {
    this._sleep_mode_outputIntialized = true;
    this._refObject.sleep_mode_output = value;
  }
  get sleep_mode_output_valid() {
    if (!this._sleep_mode_output_validIntialized) {
      return undefined;
    }
    return this._refObject.sleep_mode_output_valid;
  }
  set sleep_mode_output_valid(value) {
    this._sleep_mode_output_validIntialized = true;
    this._refObject.sleep_mode_output_valid = value;
  }
  get error_flag_contamination_warning() {
    if (!this._error_flag_contamination_warningIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_contamination_warning;
  }
  set error_flag_contamination_warning(value) {
    this._error_flag_contamination_warningIntialized = true;
    this._refObject.error_flag_contamination_warning = value;
  }
  get error_flag_contamination_error() {
    if (!this._error_flag_contamination_errorIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_contamination_error;
  }
  set error_flag_contamination_error(value) {
    this._error_flag_contamination_errorIntialized = true;
    this._refObject.error_flag_contamination_error = value;
  }
  get error_flag_manipulation_error() {
    if (!this._error_flag_manipulation_errorIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_manipulation_error;
  }
  set error_flag_manipulation_error(value) {
    this._error_flag_manipulation_errorIntialized = true;
    this._refObject.error_flag_manipulation_error = value;
  }
  get error_flag_glare() {
    if (!this._error_flag_glareIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_glare;
  }
  set error_flag_glare(value) {
    this._error_flag_glareIntialized = true;
    this._refObject.error_flag_glare = value;
  }
  get error_flag_reference_contour_intruded() {
    if (!this._error_flag_reference_contour_intrudedIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_reference_contour_intruded;
  }
  set error_flag_reference_contour_intruded(value) {
    this._error_flag_reference_contour_intrudedIntialized = true;
    this._refObject.error_flag_reference_contour_intruded = value;
  }
  get error_flag_critical_error() {
    if (!this._error_flag_critical_errorIntialized) {
      return undefined;
    }
    return this._refObject.error_flag_critical_error;
  }
  set error_flag_critical_error(value) {
    this._error_flag_critical_errorIntialized = true;
    this._refObject.error_flag_critical_error = value;
  }
  get error_flags_are_valid() {
    if (!this._error_flags_are_validIntialized) {
      return undefined;
    }
    return this._refObject.error_flags_are_valid;
  }
  set error_flags_are_valid(value) {
    this._error_flags_are_validIntialized = true;
    this._refObject.error_flags_are_valid = value;
  }
  get linear_velocity_outputs_velocity_0() {
    if (!this._linear_velocity_outputs_velocity_0Intialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_0;
  }
  set linear_velocity_outputs_velocity_0(value) {
    this._linear_velocity_outputs_velocity_0Intialized = true;
    this._refObject.linear_velocity_outputs_velocity_0 = value;
  }
  get linear_velocity_outputs_velocity_0_valid() {
    if (!this._linear_velocity_outputs_velocity_0_validIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_0_valid;
  }
  set linear_velocity_outputs_velocity_0_valid(value) {
    this._linear_velocity_outputs_velocity_0_validIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_0_valid = value;
  }
  get linear_velocity_outputs_velocity_0_transmitted_safely() {
    if (!this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_0_transmitted_safely;
  }
  set linear_velocity_outputs_velocity_0_transmitted_safely(value) {
    this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_0_transmitted_safely = value;
  }
  get linear_velocity_outputs_velocity_1() {
    if (!this._linear_velocity_outputs_velocity_1Intialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_1;
  }
  set linear_velocity_outputs_velocity_1(value) {
    this._linear_velocity_outputs_velocity_1Intialized = true;
    this._refObject.linear_velocity_outputs_velocity_1 = value;
  }
  get linear_velocity_outputs_velocity_1_valid() {
    if (!this._linear_velocity_outputs_velocity_1_validIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_1_valid;
  }
  set linear_velocity_outputs_velocity_1_valid(value) {
    this._linear_velocity_outputs_velocity_1_validIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_1_valid = value;
  }
  get linear_velocity_outputs_velocity_1_transmitted_safely() {
    if (!this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_outputs_velocity_1_transmitted_safely;
  }
  set linear_velocity_outputs_velocity_1_transmitted_safely(value) {
    this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_outputs_velocity_1_transmitted_safely = value;
  }
  get resulting_velocity() {
    return this._wrapperFields['resulting_velocity'].data;
  }
  set resulting_velocity(value) {
    this._wrapperFields['resulting_velocity'].fill(value);
  }
  get resulting_velocity_flags() {
    return this._resulting_velocity_flagsArray;
  }
  set resulting_velocity_flags(value) {
    this._resulting_velocity_flagsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ApplicationOutputsRefStruct(refObject.toObject());
    refObject.evaluation_path_outputs_eval_out.data.length = refObject.evaluation_path_outputs_eval_out.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_eval_out.size; index++) {
      this._evaluation_path_outputs_eval_outArray[index] = refObject.evaluation_path_outputs_eval_out.data[index].data;
    }
    refObject.evaluation_path_outputs_is_safe.data.length = refObject.evaluation_path_outputs_is_safe.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_is_safe.size; index++) {
      this._evaluation_path_outputs_is_safeArray[index] = refObject.evaluation_path_outputs_is_safe.data[index].data;
    }
    refObject.evaluation_path_outputs_is_valid.data.length = refObject.evaluation_path_outputs_is_valid.size;
    for (let index = 0; index < refObject.evaluation_path_outputs_is_valid.size; index++) {
      this._evaluation_path_outputs_is_validArray[index] = refObject.evaluation_path_outputs_is_valid.data[index].data;
    }
    this._wrapperFields.monitoring_case_number_outputs.copyRefObject(this._refObject.monitoring_case_number_outputs);
    refObject.monitoring_case_number_outputs_flags.data.length = refObject.monitoring_case_number_outputs_flags.size;
    for (let index = 0; index < refObject.monitoring_case_number_outputs_flags.size; index++) {
      this._monitoring_case_number_outputs_flagsArray[index] = refObject.monitoring_case_number_outputs_flags.data[index].data;
    }
    this._sleep_mode_outputIntialized = true;
    this._sleep_mode_output_validIntialized = true;
    this._error_flag_contamination_warningIntialized = true;
    this._error_flag_contamination_errorIntialized = true;
    this._error_flag_manipulation_errorIntialized = true;
    this._error_flag_glareIntialized = true;
    this._error_flag_reference_contour_intrudedIntialized = true;
    this._error_flag_critical_errorIntialized = true;
    this._error_flags_are_validIntialized = true;
    this._linear_velocity_outputs_velocity_0Intialized = true;
    this._linear_velocity_outputs_velocity_0_validIntialized = true;
    this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = true;
    this._linear_velocity_outputs_velocity_1Intialized = true;
    this._linear_velocity_outputs_velocity_1_validIntialized = true;
    this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = true;
    this._wrapperFields.resulting_velocity.copyRefObject(this._refObject.resulting_velocity);
    refObject.resulting_velocity_flags.data.length = refObject.resulting_velocity_flags.size;
    for (let index = 0; index < refObject.resulting_velocity_flags.size; index++) {
      this._resulting_velocity_flagsArray[index] = refObject.resulting_velocity_flags.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ApplicationOutputsRefStruct(other._refObject.toObject());
    this._evaluation_path_outputs_eval_outArray = other._evaluation_path_outputs_eval_outArray.slice();
    this._evaluation_path_outputs_is_safeArray = other._evaluation_path_outputs_is_safeArray.slice();
    this._evaluation_path_outputs_is_validArray = other._evaluation_path_outputs_is_validArray.slice();
    this._wrapperFields.monitoring_case_number_outputs.copy(other._wrapperFields.monitoring_case_number_outputs);
    this._monitoring_case_number_outputs_flagsArray = other._monitoring_case_number_outputs_flagsArray.slice();
    this._sleep_mode_outputIntialized = true;
    this._sleep_mode_output_validIntialized = true;
    this._error_flag_contamination_warningIntialized = true;
    this._error_flag_contamination_errorIntialized = true;
    this._error_flag_manipulation_errorIntialized = true;
    this._error_flag_glareIntialized = true;
    this._error_flag_reference_contour_intrudedIntialized = true;
    this._error_flag_critical_errorIntialized = true;
    this._error_flags_are_validIntialized = true;
    this._linear_velocity_outputs_velocity_0Intialized = true;
    this._linear_velocity_outputs_velocity_0_validIntialized = true;
    this._linear_velocity_outputs_velocity_0_transmitted_safelyIntialized = true;
    this._linear_velocity_outputs_velocity_1Intialized = true;
    this._linear_velocity_outputs_velocity_1_validIntialized = true;
    this._linear_velocity_outputs_velocity_1_transmitted_safelyIntialized = true;
    this._wrapperFields.resulting_velocity.copy(other._wrapperFields.resulting_velocity);
    this._resulting_velocity_flagsArray = other._resulting_velocity_flagsArray.slice();
  }
  static get classType() {
    return ApplicationOutputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"evaluation_path_outputs_eval_out","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"evaluation_path_outputs_is_safe","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"evaluation_path_outputs_is_valid","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"monitoring_case_number_outputs","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"monitoring_case_number_outputs_flags","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sleep_mode_output","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sleep_mode_output_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_contamination_warning","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_contamination_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_manipulation_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_glare","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_reference_contour_intruded","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flag_critical_error","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_flags_are_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_0_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_0_transmitted_safely","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_1_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_outputs_velocity_1_transmitted_safely","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resulting_velocity","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"resulting_velocity_flags","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"ApplicationOutputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ApplicationOutputs"};
  }
  hasMember(name) {
    let memberNames = ["evaluation_path_outputs_eval_out","evaluation_path_outputs_is_safe","evaluation_path_outputs_is_valid","monitoring_case_number_outputs","monitoring_case_number_outputs_flags","sleep_mode_output","sleep_mode_output_valid","error_flag_contamination_warning","error_flag_contamination_error","error_flag_manipulation_error","error_flag_glare","error_flag_reference_contour_intruded","error_flag_critical_error","error_flags_are_valid","linear_velocity_outputs_velocity_0","linear_velocity_outputs_velocity_0_valid","linear_velocity_outputs_velocity_0_transmitted_safely","linear_velocity_outputs_velocity_1","linear_velocity_outputs_velocity_1_valid","linear_velocity_outputs_velocity_1_transmitted_safely","resulting_velocity","resulting_velocity_flags"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ApplicationOutputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ApplicationOutputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ApplicationOutputsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationOutputsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationOutputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ApplicationOutputsRefArray(size);
    this._refObject = new ApplicationOutputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ApplicationOutputsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ApplicationOutputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ApplicationOutputsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ApplicationOutputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ApplicationOutputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ApplicationOutputsArrayWrapper;
  }
}
module.exports = ApplicationOutputsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "evaluation_path_outputs_eval_out",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "evaluation_path_outputs_is_safe",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "evaluation_path_outputs_is_valid",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "monitoring_case_number_outputs",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "monitoring_case_number_outputs_flags",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sleep_mode_output",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sleep_mode_output_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_contamination_warning",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_contamination_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_manipulation_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_glare",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_reference_contour_intruded",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flag_critical_error",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_flags_are_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_0",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_0_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_0_transmitted_safely",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_1_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_outputs_velocity_1_transmitted_safely",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resulting_velocity",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "resulting_velocity_flags",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "ApplicationOutputs",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ApplicationOutputs"
}
*/
