// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__DerivedValues.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const DerivedValuesRefStruct = StructType({
  multiplication_factor: primitiveTypes.uint16,
  number_of_beams: primitiveTypes.uint16,
  scan_time: primitiveTypes.uint16,
  start_angle: primitiveTypes.float32,
  angular_beam_resolution: primitiveTypes.float32,
  interbeam_period: primitiveTypes.uint32,
});
const DerivedValuesRefArray = ArrayType(DerivedValuesRefStruct);
const DerivedValuesRefStructArray = StructType({
  data: DerivedValuesRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DerivedValuesWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DerivedValuesRefStruct(other._refObject.toObject());
      this._multiplication_factorIntialized = true;
      this._number_of_beamsIntialized = true;
      this._scan_timeIntialized = true;
      this._start_angleIntialized = true;
      this._angular_beam_resolutionIntialized = true;
      this._interbeam_periodIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DerivedValuesRefStruct();
    this._multiplication_factorIntialized = false;
    this._number_of_beamsIntialized = false;
    this._scan_timeIntialized = false;
    this._start_angleIntialized = false;
    this._angular_beam_resolutionIntialized = false;
    this._interbeam_periodIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new DerivedValuesWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DerivedValuesArrayWrapper;
  }
  static get ArrayType() {
    return DerivedValuesArrayWrapper;
  }
  static get refObjectArrayType() {
    return DerivedValuesRefStructArray
  }
  static get refObjectType() {
    return DerivedValuesRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._multiplication_factorIntialized) {
        throw new TypeError('Invalid argument: multiplication_factor in DerivedValues');
      }
      if (checkConsistency && !this._number_of_beamsIntialized) {
        throw new TypeError('Invalid argument: number_of_beams in DerivedValues');
      }
      if (checkConsistency && !this._scan_timeIntialized) {
        throw new TypeError('Invalid argument: scan_time in DerivedValues');
      }
      if (checkConsistency && !this._start_angleIntialized) {
        throw new TypeError('Invalid argument: start_angle in DerivedValues');
      }
      if (checkConsistency && !this._angular_beam_resolutionIntialized) {
        throw new TypeError('Invalid argument: angular_beam_resolution in DerivedValues');
      }
      if (checkConsistency && !this._interbeam_periodIntialized) {
        throw new TypeError('Invalid argument: interbeam_period in DerivedValues');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._multiplication_factorIntialized = true;
    this._refObject.multiplication_factor = refObject.multiplication_factor;
    this._number_of_beamsIntialized = true;
    this._refObject.number_of_beams = refObject.number_of_beams;
    this._scan_timeIntialized = true;
    this._refObject.scan_time = refObject.scan_time;
    this._start_angleIntialized = true;
    this._refObject.start_angle = refObject.start_angle;
    this._angular_beam_resolutionIntialized = true;
    this._refObject.angular_beam_resolution = refObject.angular_beam_resolution;
    this._interbeam_periodIntialized = true;
    this._refObject.interbeam_period = refObject.interbeam_period;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DerivedValuesWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'DerivedValues'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get multiplication_factor() {
    if (!this._multiplication_factorIntialized) {
      return undefined;
    }
    return this._refObject.multiplication_factor;
  }
  set multiplication_factor(value) {
    this._multiplication_factorIntialized = true;
    this._refObject.multiplication_factor = value;
  }
  get number_of_beams() {
    if (!this._number_of_beamsIntialized) {
      return undefined;
    }
    return this._refObject.number_of_beams;
  }
  set number_of_beams(value) {
    this._number_of_beamsIntialized = true;
    this._refObject.number_of_beams = value;
  }
  get scan_time() {
    if (!this._scan_timeIntialized) {
      return undefined;
    }
    return this._refObject.scan_time;
  }
  set scan_time(value) {
    this._scan_timeIntialized = true;
    this._refObject.scan_time = value;
  }
  get start_angle() {
    if (!this._start_angleIntialized) {
      return undefined;
    }
    return this._refObject.start_angle;
  }
  set start_angle(value) {
    this._start_angleIntialized = true;
    this._refObject.start_angle = value;
  }
  get angular_beam_resolution() {
    if (!this._angular_beam_resolutionIntialized) {
      return undefined;
    }
    return this._refObject.angular_beam_resolution;
  }
  set angular_beam_resolution(value) {
    this._angular_beam_resolutionIntialized = true;
    this._refObject.angular_beam_resolution = value;
  }
  get interbeam_period() {
    if (!this._interbeam_periodIntialized) {
      return undefined;
    }
    return this._refObject.interbeam_period;
  }
  set interbeam_period(value) {
    this._interbeam_periodIntialized = true;
    this._refObject.interbeam_period = value;
  }
  copyRefObject(refObject) {
    this._refObject = new DerivedValuesRefStruct(refObject.toObject());
    this._multiplication_factorIntialized = true;
    this._number_of_beamsIntialized = true;
    this._scan_timeIntialized = true;
    this._start_angleIntialized = true;
    this._angular_beam_resolutionIntialized = true;
    this._interbeam_periodIntialized = true;
  }
  copy(other) {
    this._refObject = new DerivedValuesRefStruct(other._refObject.toObject());
    this._multiplication_factorIntialized = true;
    this._number_of_beamsIntialized = true;
    this._scan_timeIntialized = true;
    this._start_angleIntialized = true;
    this._angular_beam_resolutionIntialized = true;
    this._interbeam_periodIntialized = true;
  }
  static get classType() {
    return DerivedValuesWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"multiplication_factor","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"number_of_beams","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scan_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_beam_resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"interbeam_period","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"DerivedValues","stringUpperBound":null,"isPrimitiveType":false},"msgName":"DerivedValues"};
  }
  hasMember(name) {
    let memberNames = ["multiplication_factor","number_of_beams","scan_time","start_angle","angular_beam_resolution","interbeam_period"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DerivedValuesArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DerivedValuesWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DerivedValuesWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DerivedValuesArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DerivedValuesArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DerivedValuesRefArray(size);
    this._refObject = new DerivedValuesRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DerivedValuesWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DerivedValuesArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DerivedValuesArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DerivedValuesWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DerivedValuesWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DerivedValuesArrayWrapper;
  }
}
module.exports = DerivedValuesWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "multiplication_factor",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "number_of_beams",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scan_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "start_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_beam_resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "interbeam_period",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "DerivedValues",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "DerivedValues"
}
*/
