// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__MeasurementData.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const ScanPointWrapper = require('../../generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ScanPoint.js');
const MeasurementDataRefStruct = StructType({
  number_of_beams: primitiveTypes.uint32,
  scan_points: ScanPointWrapper.refObjectArrayType,
});
const MeasurementDataRefArray = ArrayType(MeasurementDataRefStruct);
const MeasurementDataRefStructArray = StructType({
  data: MeasurementDataRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MeasurementDataWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MeasurementDataRefStruct(other._refObject.toObject());
      this._number_of_beamsIntialized = true;
      this._wrapperFields.scan_points = ScanPointWrapper.createArray();
      this._wrapperFields.scan_points.copy(other._wrapperFields.scan_points);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MeasurementDataRefStruct();
    this._number_of_beamsIntialized = false;
    this._wrapperFields.scan_points = ScanPointWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new MeasurementDataWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MeasurementDataArrayWrapper;
  }
  static get ArrayType() {
    return MeasurementDataArrayWrapper;
  }
  static get refObjectArrayType() {
    return MeasurementDataRefStructArray
  }
  static get refObjectType() {
    return MeasurementDataRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._number_of_beamsIntialized) {
        throw new TypeError('Invalid argument: number_of_beams in MeasurementData');
      }
    this._wrapperFields.scan_points.freeze(own, checkConsistency);
    this._refObject.scan_points = this._wrapperFields.scan_points.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._number_of_beamsIntialized = true;
    this._refObject.number_of_beams = refObject.number_of_beams;
    this._wrapperFields.scan_points.copyRefObject(refObject.scan_points);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.scan_points.size != 0) {
      ScanPointWrapper.ArrayType.freeArray(refObject.scan_points);
      if (ScanPointWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.scan_points, ScanPointWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    MeasurementDataWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'MeasurementData'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get number_of_beams() {
    if (!this._number_of_beamsIntialized) {
      return undefined;
    }
    return this._refObject.number_of_beams;
  }
  set number_of_beams(value) {
    this._number_of_beamsIntialized = true;
    this._refObject.number_of_beams = value;
  }
  get scan_points() {
    return this._wrapperFields.scan_points;
  }
  set scan_points(value) {
    this._wrapperFields.scan_points.fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new MeasurementDataRefStruct(refObject.toObject());
    this._number_of_beamsIntialized = true;
    this._wrapperFields.scan_points.copyRefObject(this._refObject.scan_points);
  }
  copy(other) {
    this._refObject = new MeasurementDataRefStruct(other._refObject.toObject());
    this._number_of_beamsIntialized = true;
    this._wrapperFields.scan_points.copy(other._wrapperFields.scan_points);
  }
  static get classType() {
    return MeasurementDataWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"number_of_beams","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"scan_points","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"sick_safetyscanners2_interfaces","type":"ScanPoint","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"MeasurementData","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MeasurementData"};
  }
  hasMember(name) {
    let memberNames = ["number_of_beams","scan_points"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MeasurementDataArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MeasurementDataWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MeasurementDataWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MeasurementDataArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MeasurementDataArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MeasurementDataRefArray(size);
    this._refObject = new MeasurementDataRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MeasurementDataWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MeasurementDataArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MeasurementDataArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MeasurementDataWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MeasurementDataWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MeasurementDataArrayWrapper;
  }
}
module.exports = MeasurementDataWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "number_of_beams",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "scan_points",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "sick_safetyscanners2_interfaces",
        "type": "ScanPoint",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "MeasurementData",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MeasurementData"
}
*/
