// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__msg__Link.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const TransformWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Transform.js');
const Float64Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float64.js');
const LinkRefStruct = StructType({
  from_id: primitiveTypes.int32,
  to_id: primitiveTypes.int32,
  type: primitiveTypes.int32,
  transform: TransformWrapper.refObjectType,
  information: ArrayType(primitiveTypes.float64, 36),
});
const LinkRefArray = ArrayType(LinkRefStruct);
const LinkRefStructArray = StructType({
  data: LinkRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LinkWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LinkRefStruct(other._refObject.toObject());
      this._from_idIntialized = true;
      this._to_idIntialized = true;
      this._typeIntialized = true;
      this._wrapperFields.transform =  new TransformWrapper(other._wrapperFields.transform);
      this._wrapperFields.information = Float64Wrapper.createArray();
      this._wrapperFields.information.copy(other._wrapperFields.information);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LinkRefStruct();
    this._from_idIntialized = false;
    this._to_idIntialized = false;
    this._typeIntialized = false;
    this._wrapperFields.transform = new TransformWrapper();
    this._wrapperFields.information = Float64Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new LinkWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LinkArrayWrapper;
  }
  static get ArrayType() {
    return LinkArrayWrapper;
  }
  static get refObjectArrayType() {
    return LinkRefStructArray
  }
  static get refObjectType() {
    return LinkRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._from_idIntialized) {
        throw new TypeError('Invalid argument: from_id in Link');
      }
      if (checkConsistency && !this._to_idIntialized) {
        throw new TypeError('Invalid argument: to_id in Link');
      }
      if (checkConsistency && !this._typeIntialized) {
        throw new TypeError('Invalid argument: type in Link');
      }
    this._wrapperFields.transform.freeze(own, checkConsistency);
    this._refObject.transform = this._wrapperFields.transform.refObject;
    for (let i = 0; i < this._wrapperFields.information.data.length; i++) {
      this._refObject.information[i] = this._wrapperFields.information.data[i];
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._from_idIntialized = true;
    this._refObject.from_id = refObject.from_id;
    this._to_idIntialized = true;
    this._refObject.to_id = refObject.to_id;
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._wrapperFields.transform.copyRefObject(refObject.transform);
    this._refObject.information = refObject.information;
    this._wrapperFields.information.fill(refObject.information.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    TransformWrapper.freeStruct(refObject.transform);
  }
  static destoryRawROS(msg) {
    LinkWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'msg', interfaceName: 'Link'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get from_id() {
    if (!this._from_idIntialized) {
      return undefined;
    }
    return this._refObject.from_id;
  }
  set from_id(value) {
    this._from_idIntialized = true;
    this._refObject.from_id = value;
  }
  get to_id() {
    if (!this._to_idIntialized) {
      return undefined;
    }
    return this._refObject.to_id;
  }
  set to_id(value) {
    this._to_idIntialized = true;
    this._refObject.to_id = value;
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get transform() {
    return this._wrapperFields.transform;
  }
  set transform(value) {
    if (value instanceof TransformWrapper) {
      this._wrapperFields.transform.copy(value);
    } else {
      this._wrapperFields.transform.copy(new TransformWrapper(value));
    }
  }
  get information() {
    return this._wrapperFields['information'].data;
  }
  set information(value) {
    if (value.length !== 36) {
      throw new RangeError('The length of the array must be 36.');
    }
    this._wrapperFields['information'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new LinkRefStruct(refObject.toObject());
    this._from_idIntialized = true;
    this._to_idIntialized = true;
    this._typeIntialized = true;
    this._wrapperFields.transform.copyRefObject(this._refObject.transform);
    this._wrapperFields.information.fill(refObject.information.toArray());
  }
  copy(other) {
    this._refObject = new LinkRefStruct(other._refObject.toObject());
    this._from_idIntialized = true;
    this._to_idIntialized = true;
    this._typeIntialized = true;
    this._wrapperFields.transform.copy(other._wrapperFields.transform);
    this._wrapperFields.information.copy(other._wrapperFields.information);
  }
  static get classType() {
    return LinkWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"from_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"to_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"transform","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"Transform","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"information","type":{"isArray":true,"arraySize":36,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"Link","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Link"};
  }
  hasMember(name) {
    let memberNames = ["from_id","to_id","type","transform","information"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LinkArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LinkWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LinkWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LinkArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LinkRefArray(size);
    this._refObject = new LinkRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LinkWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LinkArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LinkArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LinkWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LinkWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LinkArrayWrapper;
  }
}
module.exports = LinkWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "from_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "to_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "transform",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "Transform",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "information",
      "type": {
        "isArray": true,
        "arraySize": 36,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "Link",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Link"
}
*/
