// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__ListLabels_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ListLabels_ResponseRefStruct = StructType({
  ids: Int32Wrapper.refObjectArrayType,
  labels: StringWrapper.refObjectArrayType,
});
const ListLabels_ResponseRefArray = ArrayType(ListLabels_ResponseRefStruct);
const ListLabels_ResponseRefStructArray = StructType({
  data: ListLabels_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ListLabels_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._labelsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ListLabels_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.ids = Int32Wrapper.createArray();
      this._wrapperFields.ids.copy(other._wrapperFields.ids);
      this._wrapperFields.labels = StringWrapper.createArray();
      this._wrapperFields.labels.copy(other._wrapperFields.labels);
      this.labels = other.labels;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ListLabels_ResponseRefStruct();
    this._wrapperFields.ids = Int32Wrapper.createArray();
    this._wrapperFields.labels = StringWrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new ListLabels_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ListLabels_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return ListLabels_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return ListLabels_ResponseRefStructArray
  }
  static get refObjectType() {
    return ListLabels_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.ids.freeze(own, checkConsistency);
    this._refObject.ids = this._wrapperFields.ids.refObject;
    if (own) {
      this._wrapperFields.ids.fill(Int32Array.from([]));
      this._wrapperFields.ids.freeze(own, checkConsistency);
      this._refObject.ids = this._wrapperFields.ids.refObject;
    }
    if (!own) {
      this._wrapperFields.labels.fill(this._labelsArray);
      this._wrapperFields.labels.freeze(own, checkConsistency);
      this._refObject.labels = this._wrapperFields.labels.refObject;
    } else {
      this._wrapperFields.labels.fill([]);
      this._wrapperFields.labels.freeze(own, checkConsistency);
      this._refObject.labels = this._wrapperFields.labels.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ids.copyRefObject(refObject.ids);
    refObject.labels.data.length = refObject.labels.size;
    for (let index = 0; index < refObject.labels.size; index++) {
      this._labelsArray[index] = refObject.labels.data[index].data;
    }
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ids.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.ids);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ids, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.labels.size != 0) {
      StringWrapper.ArrayType.freeArray(refObject.labels);
      if (StringWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.labels, StringWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ListLabels_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'ListLabels_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ids() {
    return this._wrapperFields['ids'].data;
  }
  set ids(value) {
    this._wrapperFields['ids'].fill(value);
  }
  get labels() {
    return this._labelsArray;
  }
  set labels(value) {
    this._labelsArray = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ListLabels_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.ids.copyRefObject(this._refObject.ids);
    refObject.labels.data.length = refObject.labels.size;
    for (let index = 0; index < refObject.labels.size; index++) {
      this._labelsArray[index] = refObject.labels.data[index].data;
    }
  }
  copy(other) {
    this._refObject = new ListLabels_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.ids.copy(other._wrapperFields.ids);
    this._labelsArray = other._labelsArray.slice();
  }
  static get classType() {
    return ListLabels_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"labels","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"ListLabels_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ListLabels_Response"};
  }
  hasMember(name) {
    let memberNames = ["ids","labels"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ListLabels_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ListLabels_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ListLabels_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListLabels_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ListLabels_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ListLabels_ResponseRefArray(size);
    this._refObject = new ListLabels_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ListLabels_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ListLabels_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ListLabels_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ListLabels_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ListLabels_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ListLabels_ResponseArrayWrapper;
  }
}
module.exports = ListLabels_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "labels",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "ListLabels_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ListLabels_Response"
}
*/
