// rclnodejs/generated/ouster_msgs/ouster_msgs__srv__GetMetadata_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MetadataWrapper = require('../../generated/ouster_msgs/ouster_msgs__msg__Metadata.js');
const GetMetadata_ResponseRefStruct = StructType({
  metadata: MetadataWrapper.refObjectType,
});
const GetMetadata_ResponseRefArray = ArrayType(GetMetadata_ResponseRefStruct);
const GetMetadata_ResponseRefStructArray = StructType({
  data: GetMetadata_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class GetMetadata_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new GetMetadata_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.metadata =  new MetadataWrapper(other._wrapperFields.metadata);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new GetMetadata_ResponseRefStruct();
    this._wrapperFields.metadata = new MetadataWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new GetMetadata_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new GetMetadata_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return GetMetadata_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return GetMetadata_ResponseRefStructArray
  }
  static get refObjectType() {
    return GetMetadata_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.metadata.freeze(own, checkConsistency);
    this._refObject.metadata = this._wrapperFields.metadata.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.metadata.copyRefObject(refObject.metadata);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    MetadataWrapper.freeStruct(refObject.metadata);
  }
  static destoryRawROS(msg) {
    GetMetadata_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'ouster_msgs', subFolder: 'srv', interfaceName: 'GetMetadata_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get metadata() {
    return this._wrapperFields.metadata;
  }
  set metadata(value) {
    if (value instanceof MetadataWrapper) {
      this._wrapperFields.metadata.copy(value);
    } else {
      this._wrapperFields.metadata.copy(new MetadataWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new GetMetadata_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.metadata.copyRefObject(this._refObject.metadata);
  }
  copy(other) {
    this._refObject = new GetMetadata_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.metadata.copy(other._wrapperFields.metadata);
  }
  static get classType() {
    return GetMetadata_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"metadata","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"ouster_msgs","type":"Metadata","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"ouster_msgs","type":"GetMetadata_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"GetMetadata_Response"};
  }
  hasMember(name) {
    let memberNames = ["metadata"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class GetMetadata_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof GetMetadata_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new GetMetadata_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMetadata_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to GetMetadata_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new GetMetadata_ResponseRefArray(size);
    this._refObject = new GetMetadata_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new GetMetadata_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof GetMetadata_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "GetMetadata_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      GetMetadata_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return GetMetadata_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return GetMetadata_ResponseArrayWrapper;
  }
}
module.exports = GetMetadata_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "metadata",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "ouster_msgs",
        "type": "Metadata",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "ouster_msgs",
    "type": "GetMetadata_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "GetMetadata_Response"
}
*/
