// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__Field.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const FieldRefStruct = StructType({
  ranges: Float32Wrapper.refObjectArrayType,
  start_angle: primitiveTypes.float32,
  angular_resolution: primitiveTypes.float32,
  protective_field: primitiveTypes.bool,
});
const FieldRefArray = ArrayType(FieldRefStruct);
const FieldRefStructArray = StructType({
  data: FieldRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FieldWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FieldRefStruct(other._refObject.toObject());
      this._wrapperFields.ranges = Float32Wrapper.createArray();
      this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
      this._start_angleIntialized = true;
      this._angular_resolutionIntialized = true;
      this._protective_fieldIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FieldRefStruct();
    this._wrapperFields.ranges = Float32Wrapper.createArray();
    this._start_angleIntialized = false;
    this._angular_resolutionIntialized = false;
    this._protective_fieldIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FieldWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FieldArrayWrapper;
  }
  static get ArrayType() {
    return FieldArrayWrapper;
  }
  static get refObjectArrayType() {
    return FieldRefStructArray
  }
  static get refObjectType() {
    return FieldRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._start_angleIntialized) {
        throw new TypeError('Invalid argument: start_angle in Field');
      }
      if (checkConsistency && !this._angular_resolutionIntialized) {
        throw new TypeError('Invalid argument: angular_resolution in Field');
      }
      if (checkConsistency && !this._protective_fieldIntialized) {
        throw new TypeError('Invalid argument: protective_field in Field');
      }
    this._wrapperFields.ranges.freeze(own, checkConsistency);
    this._refObject.ranges = this._wrapperFields.ranges.refObject;
    if (own) {
      this._wrapperFields.ranges.fill(Float32Array.from([]));
      this._wrapperFields.ranges.freeze(own, checkConsistency);
      this._refObject.ranges = this._wrapperFields.ranges.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.ranges.copyRefObject(refObject.ranges);
    this._start_angleIntialized = true;
    this._refObject.start_angle = refObject.start_angle;
    this._angular_resolutionIntialized = true;
    this._refObject.angular_resolution = refObject.angular_resolution;
    this._protective_fieldIntialized = true;
    this._refObject.protective_field = refObject.protective_field;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.ranges.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.ranges);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.ranges, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    FieldWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'Field'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get ranges() {
    return this._wrapperFields['ranges'].data;
  }
  set ranges(value) {
    this._wrapperFields['ranges'].fill(value);
  }
  get start_angle() {
    if (!this._start_angleIntialized) {
      return undefined;
    }
    return this._refObject.start_angle;
  }
  set start_angle(value) {
    this._start_angleIntialized = true;
    this._refObject.start_angle = value;
  }
  get angular_resolution() {
    if (!this._angular_resolutionIntialized) {
      return undefined;
    }
    return this._refObject.angular_resolution;
  }
  set angular_resolution(value) {
    this._angular_resolutionIntialized = true;
    this._refObject.angular_resolution = value;
  }
  get protective_field() {
    if (!this._protective_fieldIntialized) {
      return undefined;
    }
    return this._refObject.protective_field;
  }
  set protective_field(value) {
    this._protective_fieldIntialized = true;
    this._refObject.protective_field = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FieldRefStruct(refObject.toObject());
    this._wrapperFields.ranges.copyRefObject(this._refObject.ranges);
    this._start_angleIntialized = true;
    this._angular_resolutionIntialized = true;
    this._protective_fieldIntialized = true;
  }
  copy(other) {
    this._refObject = new FieldRefStruct(other._refObject.toObject());
    this._wrapperFields.ranges.copy(other._wrapperFields.ranges);
    this._start_angleIntialized = true;
    this._angular_resolutionIntialized = true;
    this._protective_fieldIntialized = true;
  }
  static get classType() {
    return FieldWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"ranges","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"start_angle","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"angular_resolution","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"protective_field","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"Field","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Field"};
  }
  hasMember(name) {
    let memberNames = ["ranges","start_angle","angular_resolution","protective_field"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FieldArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FieldWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FieldWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FieldArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FieldRefArray(size);
    this._refObject = new FieldRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FieldWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FieldArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FieldArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FieldWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FieldWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FieldArrayWrapper;
  }
}
module.exports = FieldWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "ranges",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "start_angle",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "angular_resolution",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "protective_field",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "Field",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Field"
}
*/
