// rclnodejs/generated/sick_safetyscanners2_interfaces/sick_safetyscanners2_interfaces__msg__ApplicationInputs.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const BoolWrapper = require('../../generated/std_msgs/std_msgs__msg__Bool.js');
const UInt16Wrapper = require('../../generated/std_msgs/std_msgs__msg__UInt16.js');
const ApplicationInputsRefStruct = StructType({
  unsafe_inputs_input_sources: BoolWrapper.refObjectArrayType,
  unsafe_inputs_flags: BoolWrapper.refObjectArrayType,
  monitoring_case_number_inputs: UInt16Wrapper.refObjectArrayType,
  monitoring_case_number_inputs_flags: BoolWrapper.refObjectArrayType,
  linear_velocity_inputs_velocity_0: primitiveTypes.int16,
  linear_velocity_inputs_velocity_0_valid: primitiveTypes.bool,
  linear_velocity_inputs_velocity_0_transmitted_safely: primitiveTypes.bool,
  linear_velocity_inputs_velocity_1: primitiveTypes.int16,
  linear_velocity_inputs_velocity_1_valid: primitiveTypes.bool,
  linear_velocity_inputs_velocity_1_transmitted_safely: primitiveTypes.bool,
  sleep_mode_input: primitiveTypes.uint8,
});
const ApplicationInputsRefArray = ArrayType(ApplicationInputsRefStruct);
const ApplicationInputsRefStructArray = StructType({
  data: ApplicationInputsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ApplicationInputsWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    this._unsafe_inputs_input_sourcesArray = [];
    this._unsafe_inputs_flagsArray = [];
    this._monitoring_case_number_inputs_flagsArray = [];
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ApplicationInputsRefStruct(other._refObject.toObject());
      this._wrapperFields.unsafe_inputs_input_sources = BoolWrapper.createArray();
      this._wrapperFields.unsafe_inputs_input_sources.copy(other._wrapperFields.unsafe_inputs_input_sources);
      this.unsafe_inputs_input_sources = other.unsafe_inputs_input_sources;
      this._wrapperFields.unsafe_inputs_flags = BoolWrapper.createArray();
      this._wrapperFields.unsafe_inputs_flags.copy(other._wrapperFields.unsafe_inputs_flags);
      this.unsafe_inputs_flags = other.unsafe_inputs_flags;
      this._wrapperFields.monitoring_case_number_inputs = UInt16Wrapper.createArray();
      this._wrapperFields.monitoring_case_number_inputs.copy(other._wrapperFields.monitoring_case_number_inputs);
      this._wrapperFields.monitoring_case_number_inputs_flags = BoolWrapper.createArray();
      this._wrapperFields.monitoring_case_number_inputs_flags.copy(other._wrapperFields.monitoring_case_number_inputs_flags);
      this.monitoring_case_number_inputs_flags = other.monitoring_case_number_inputs_flags;
      this._linear_velocity_inputs_velocity_0Intialized = true;
      this._linear_velocity_inputs_velocity_0_validIntialized = true;
      this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = true;
      this._linear_velocity_inputs_velocity_1Intialized = true;
      this._linear_velocity_inputs_velocity_1_validIntialized = true;
      this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = true;
      this._sleep_mode_inputIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ApplicationInputsRefStruct();
    this._wrapperFields.unsafe_inputs_input_sources = BoolWrapper.createArray();
    this._wrapperFields.unsafe_inputs_flags = BoolWrapper.createArray();
    this._wrapperFields.monitoring_case_number_inputs = UInt16Wrapper.createArray();
    this._wrapperFields.monitoring_case_number_inputs_flags = BoolWrapper.createArray();
    this._linear_velocity_inputs_velocity_0Intialized = false;
    this._linear_velocity_inputs_velocity_0_validIntialized = false;
    this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = false;
    this._linear_velocity_inputs_velocity_1Intialized = false;
    this._linear_velocity_inputs_velocity_1_validIntialized = false;
    this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = false;
    this._sleep_mode_inputIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ApplicationInputsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ApplicationInputsArrayWrapper;
  }
  static get ArrayType() {
    return ApplicationInputsArrayWrapper;
  }
  static get refObjectArrayType() {
    return ApplicationInputsRefStructArray
  }
  static get refObjectType() {
    return ApplicationInputsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._linear_velocity_inputs_velocity_0Intialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_0 in ApplicationInputs');
      }
      if (checkConsistency && !this._linear_velocity_inputs_velocity_0_validIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_0_valid in ApplicationInputs');
      }
      if (checkConsistency && !this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_0_transmitted_safely in ApplicationInputs');
      }
      if (checkConsistency && !this._linear_velocity_inputs_velocity_1Intialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_1 in ApplicationInputs');
      }
      if (checkConsistency && !this._linear_velocity_inputs_velocity_1_validIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_1_valid in ApplicationInputs');
      }
      if (checkConsistency && !this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized) {
        throw new TypeError('Invalid argument: linear_velocity_inputs_velocity_1_transmitted_safely in ApplicationInputs');
      }
      if (checkConsistency && !this._sleep_mode_inputIntialized) {
        throw new TypeError('Invalid argument: sleep_mode_input in ApplicationInputs');
      }
    if (!own) {
      this._wrapperFields.unsafe_inputs_input_sources.fill(this._unsafe_inputs_input_sourcesArray);
      this._wrapperFields.unsafe_inputs_input_sources.freeze(own, checkConsistency);
      this._refObject.unsafe_inputs_input_sources = this._wrapperFields.unsafe_inputs_input_sources.refObject;
    } else {
      this._wrapperFields.unsafe_inputs_input_sources.fill([]);
      this._wrapperFields.unsafe_inputs_input_sources.freeze(own, checkConsistency);
      this._refObject.unsafe_inputs_input_sources = this._wrapperFields.unsafe_inputs_input_sources.refObject;
    }
    if (!own) {
      this._wrapperFields.unsafe_inputs_flags.fill(this._unsafe_inputs_flagsArray);
      this._wrapperFields.unsafe_inputs_flags.freeze(own, checkConsistency);
      this._refObject.unsafe_inputs_flags = this._wrapperFields.unsafe_inputs_flags.refObject;
    } else {
      this._wrapperFields.unsafe_inputs_flags.fill([]);
      this._wrapperFields.unsafe_inputs_flags.freeze(own, checkConsistency);
      this._refObject.unsafe_inputs_flags = this._wrapperFields.unsafe_inputs_flags.refObject;
    }
    this._wrapperFields.monitoring_case_number_inputs.freeze(own, checkConsistency);
    this._refObject.monitoring_case_number_inputs = this._wrapperFields.monitoring_case_number_inputs.refObject;
    if (own) {
      this._wrapperFields.monitoring_case_number_inputs.fill(Uint16Array.from([]));
      this._wrapperFields.monitoring_case_number_inputs.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_inputs = this._wrapperFields.monitoring_case_number_inputs.refObject;
    }
    if (!own) {
      this._wrapperFields.monitoring_case_number_inputs_flags.fill(this._monitoring_case_number_inputs_flagsArray);
      this._wrapperFields.monitoring_case_number_inputs_flags.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_inputs_flags = this._wrapperFields.monitoring_case_number_inputs_flags.refObject;
    } else {
      this._wrapperFields.monitoring_case_number_inputs_flags.fill([]);
      this._wrapperFields.monitoring_case_number_inputs_flags.freeze(own, checkConsistency);
      this._refObject.monitoring_case_number_inputs_flags = this._wrapperFields.monitoring_case_number_inputs_flags.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    refObject.unsafe_inputs_input_sources.data.length = refObject.unsafe_inputs_input_sources.size;
    for (let index = 0; index < refObject.unsafe_inputs_input_sources.size; index++) {
      this._unsafe_inputs_input_sourcesArray[index] = refObject.unsafe_inputs_input_sources.data[index].data;
    }
    refObject.unsafe_inputs_flags.data.length = refObject.unsafe_inputs_flags.size;
    for (let index = 0; index < refObject.unsafe_inputs_flags.size; index++) {
      this._unsafe_inputs_flagsArray[index] = refObject.unsafe_inputs_flags.data[index].data;
    }
    this._wrapperFields.monitoring_case_number_inputs.copyRefObject(refObject.monitoring_case_number_inputs);
    refObject.monitoring_case_number_inputs_flags.data.length = refObject.monitoring_case_number_inputs_flags.size;
    for (let index = 0; index < refObject.monitoring_case_number_inputs_flags.size; index++) {
      this._monitoring_case_number_inputs_flagsArray[index] = refObject.monitoring_case_number_inputs_flags.data[index].data;
    }
    this._linear_velocity_inputs_velocity_0Intialized = true;
    this._refObject.linear_velocity_inputs_velocity_0 = refObject.linear_velocity_inputs_velocity_0;
    this._linear_velocity_inputs_velocity_0_validIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_0_valid = refObject.linear_velocity_inputs_velocity_0_valid;
    this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_0_transmitted_safely = refObject.linear_velocity_inputs_velocity_0_transmitted_safely;
    this._linear_velocity_inputs_velocity_1Intialized = true;
    this._refObject.linear_velocity_inputs_velocity_1 = refObject.linear_velocity_inputs_velocity_1;
    this._linear_velocity_inputs_velocity_1_validIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_1_valid = refObject.linear_velocity_inputs_velocity_1_valid;
    this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_1_transmitted_safely = refObject.linear_velocity_inputs_velocity_1_transmitted_safely;
    this._sleep_mode_inputIntialized = true;
    this._refObject.sleep_mode_input = refObject.sleep_mode_input;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.unsafe_inputs_input_sources.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.unsafe_inputs_input_sources);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.unsafe_inputs_input_sources, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.unsafe_inputs_flags.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.unsafe_inputs_flags);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.unsafe_inputs_flags, BoolWrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.monitoring_case_number_inputs.size != 0) {
      UInt16Wrapper.ArrayType.freeArray(refObject.monitoring_case_number_inputs);
      if (UInt16Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.monitoring_case_number_inputs, UInt16Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.monitoring_case_number_inputs_flags.size != 0) {
      BoolWrapper.ArrayType.freeArray(refObject.monitoring_case_number_inputs_flags);
      if (BoolWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.monitoring_case_number_inputs_flags, BoolWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    ApplicationInputsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'sick_safetyscanners2_interfaces', subFolder: 'msg', interfaceName: 'ApplicationInputs'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get unsafe_inputs_input_sources() {
    return this._unsafe_inputs_input_sourcesArray;
  }
  set unsafe_inputs_input_sources(value) {
    this._unsafe_inputs_input_sourcesArray = value;
  }
  get unsafe_inputs_flags() {
    return this._unsafe_inputs_flagsArray;
  }
  set unsafe_inputs_flags(value) {
    this._unsafe_inputs_flagsArray = value;
  }
  get monitoring_case_number_inputs() {
    return this._wrapperFields['monitoring_case_number_inputs'].data;
  }
  set monitoring_case_number_inputs(value) {
    this._wrapperFields['monitoring_case_number_inputs'].fill(value);
  }
  get monitoring_case_number_inputs_flags() {
    return this._monitoring_case_number_inputs_flagsArray;
  }
  set monitoring_case_number_inputs_flags(value) {
    this._monitoring_case_number_inputs_flagsArray = value;
  }
  get linear_velocity_inputs_velocity_0() {
    if (!this._linear_velocity_inputs_velocity_0Intialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_0;
  }
  set linear_velocity_inputs_velocity_0(value) {
    this._linear_velocity_inputs_velocity_0Intialized = true;
    this._refObject.linear_velocity_inputs_velocity_0 = value;
  }
  get linear_velocity_inputs_velocity_0_valid() {
    if (!this._linear_velocity_inputs_velocity_0_validIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_0_valid;
  }
  set linear_velocity_inputs_velocity_0_valid(value) {
    this._linear_velocity_inputs_velocity_0_validIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_0_valid = value;
  }
  get linear_velocity_inputs_velocity_0_transmitted_safely() {
    if (!this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_0_transmitted_safely;
  }
  set linear_velocity_inputs_velocity_0_transmitted_safely(value) {
    this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_0_transmitted_safely = value;
  }
  get linear_velocity_inputs_velocity_1() {
    if (!this._linear_velocity_inputs_velocity_1Intialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_1;
  }
  set linear_velocity_inputs_velocity_1(value) {
    this._linear_velocity_inputs_velocity_1Intialized = true;
    this._refObject.linear_velocity_inputs_velocity_1 = value;
  }
  get linear_velocity_inputs_velocity_1_valid() {
    if (!this._linear_velocity_inputs_velocity_1_validIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_1_valid;
  }
  set linear_velocity_inputs_velocity_1_valid(value) {
    this._linear_velocity_inputs_velocity_1_validIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_1_valid = value;
  }
  get linear_velocity_inputs_velocity_1_transmitted_safely() {
    if (!this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized) {
      return undefined;
    }
    return this._refObject.linear_velocity_inputs_velocity_1_transmitted_safely;
  }
  set linear_velocity_inputs_velocity_1_transmitted_safely(value) {
    this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = true;
    this._refObject.linear_velocity_inputs_velocity_1_transmitted_safely = value;
  }
  get sleep_mode_input() {
    if (!this._sleep_mode_inputIntialized) {
      return undefined;
    }
    return this._refObject.sleep_mode_input;
  }
  set sleep_mode_input(value) {
    this._sleep_mode_inputIntialized = true;
    this._refObject.sleep_mode_input = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ApplicationInputsRefStruct(refObject.toObject());
    refObject.unsafe_inputs_input_sources.data.length = refObject.unsafe_inputs_input_sources.size;
    for (let index = 0; index < refObject.unsafe_inputs_input_sources.size; index++) {
      this._unsafe_inputs_input_sourcesArray[index] = refObject.unsafe_inputs_input_sources.data[index].data;
    }
    refObject.unsafe_inputs_flags.data.length = refObject.unsafe_inputs_flags.size;
    for (let index = 0; index < refObject.unsafe_inputs_flags.size; index++) {
      this._unsafe_inputs_flagsArray[index] = refObject.unsafe_inputs_flags.data[index].data;
    }
    this._wrapperFields.monitoring_case_number_inputs.copyRefObject(this._refObject.monitoring_case_number_inputs);
    refObject.monitoring_case_number_inputs_flags.data.length = refObject.monitoring_case_number_inputs_flags.size;
    for (let index = 0; index < refObject.monitoring_case_number_inputs_flags.size; index++) {
      this._monitoring_case_number_inputs_flagsArray[index] = refObject.monitoring_case_number_inputs_flags.data[index].data;
    }
    this._linear_velocity_inputs_velocity_0Intialized = true;
    this._linear_velocity_inputs_velocity_0_validIntialized = true;
    this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = true;
    this._linear_velocity_inputs_velocity_1Intialized = true;
    this._linear_velocity_inputs_velocity_1_validIntialized = true;
    this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = true;
    this._sleep_mode_inputIntialized = true;
  }
  copy(other) {
    this._refObject = new ApplicationInputsRefStruct(other._refObject.toObject());
    this._unsafe_inputs_input_sourcesArray = other._unsafe_inputs_input_sourcesArray.slice();
    this._unsafe_inputs_flagsArray = other._unsafe_inputs_flagsArray.slice();
    this._wrapperFields.monitoring_case_number_inputs.copy(other._wrapperFields.monitoring_case_number_inputs);
    this._monitoring_case_number_inputs_flagsArray = other._monitoring_case_number_inputs_flagsArray.slice();
    this._linear_velocity_inputs_velocity_0Intialized = true;
    this._linear_velocity_inputs_velocity_0_validIntialized = true;
    this._linear_velocity_inputs_velocity_0_transmitted_safelyIntialized = true;
    this._linear_velocity_inputs_velocity_1Intialized = true;
    this._linear_velocity_inputs_velocity_1_validIntialized = true;
    this._linear_velocity_inputs_velocity_1_transmitted_safelyIntialized = true;
    this._sleep_mode_inputIntialized = true;
  }
  static get classType() {
    return ApplicationInputsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"unsafe_inputs_input_sources","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"unsafe_inputs_flags","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"monitoring_case_number_inputs","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"uint16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"monitoring_case_number_inputs_flags","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_0","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_0_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_0_transmitted_safely","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_1","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_1_valid","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"linear_velocity_inputs_velocity_1_transmitted_safely","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"sleep_mode_input","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"sick_safetyscanners2_interfaces","type":"ApplicationInputs","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ApplicationInputs"};
  }
  hasMember(name) {
    let memberNames = ["unsafe_inputs_input_sources","unsafe_inputs_flags","monitoring_case_number_inputs","monitoring_case_number_inputs_flags","linear_velocity_inputs_velocity_0","linear_velocity_inputs_velocity_0_valid","linear_velocity_inputs_velocity_0_transmitted_safely","linear_velocity_inputs_velocity_1","linear_velocity_inputs_velocity_1_valid","linear_velocity_inputs_velocity_1_transmitted_safely","sleep_mode_input"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ApplicationInputsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ApplicationInputsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ApplicationInputsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationInputsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ApplicationInputsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ApplicationInputsRefArray(size);
    this._refObject = new ApplicationInputsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ApplicationInputsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ApplicationInputsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ApplicationInputsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ApplicationInputsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ApplicationInputsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ApplicationInputsArrayWrapper;
  }
}
module.exports = ApplicationInputsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "unsafe_inputs_input_sources",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "unsafe_inputs_flags",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "monitoring_case_number_inputs",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "uint16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "monitoring_case_number_inputs_flags",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_0",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_0_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_0_transmitted_safely",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_1",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_1_valid",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "linear_velocity_inputs_velocity_1_transmitted_safely",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "sleep_mode_input",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "sick_safetyscanners2_interfaces",
    "type": "ApplicationInputs",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ApplicationInputs"
}
*/
