// rclnodejs/generated/teleop_tools_msgs/teleop_tools_msgs__action__Increment_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const Increment_GoalRefStruct = StructType({
  increment_by: Float32Wrapper.refObjectArrayType,
});
const Increment_GoalRefArray = ArrayType(Increment_GoalRefStruct);
const Increment_GoalRefStructArray = StructType({
  data: Increment_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class Increment_GoalWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new Increment_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.increment_by = Float32Wrapper.createArray();
      this._wrapperFields.increment_by.copy(other._wrapperFields.increment_by);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new Increment_GoalRefStruct();
    this._wrapperFields.increment_by = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new Increment_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new Increment_GoalArrayWrapper;
  }
  static get ArrayType() {
    return Increment_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return Increment_GoalRefStructArray
  }
  static get refObjectType() {
    return Increment_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    this._wrapperFields.increment_by.freeze(own, checkConsistency);
    this._refObject.increment_by = this._wrapperFields.increment_by.refObject;
    if (own) {
      this._wrapperFields.increment_by.fill(Float32Array.from([]));
      this._wrapperFields.increment_by.freeze(own, checkConsistency);
      this._refObject.increment_by = this._wrapperFields.increment_by.refObject;
    }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.increment_by.copyRefObject(refObject.increment_by);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.increment_by.size != 0) {
      Float32Wrapper.ArrayType.freeArray(refObject.increment_by);
      if (Float32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.increment_by, Float32Wrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    Increment_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'teleop_tools_msgs', subFolder: 'action', interfaceName: 'Increment_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get increment_by() {
    return this._wrapperFields['increment_by'].data;
  }
  set increment_by(value) {
    this._wrapperFields['increment_by'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new Increment_GoalRefStruct(refObject.toObject());
    this._wrapperFields.increment_by.copyRefObject(this._refObject.increment_by);
  }
  copy(other) {
    this._refObject = new Increment_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.increment_by.copy(other._wrapperFields.increment_by);
  }
  static get classType() {
    return Increment_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"increment_by","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"teleop_tools_msgs","type":"Increment_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Increment_Goal"};
  }
  hasMember(name) {
    let memberNames = ["increment_by"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class Increment_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof Increment_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new Increment_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Increment_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to Increment_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new Increment_GoalRefArray(size);
    this._refObject = new Increment_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new Increment_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof Increment_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "Increment_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      Increment_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return Increment_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return Increment_GoalArrayWrapper;
  }
}
module.exports = Increment_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "increment_by",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "teleop_tools_msgs",
    "type": "Increment_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Increment_Goal"
}
*/
