// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__SetGoal_Response.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Int32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Int32.js');
const PoseWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__Pose.js');
const SetGoal_ResponseRefStruct = StructType({
  path_ids: Int32Wrapper.refObjectArrayType,
  path_poses: PoseWrapper.refObjectArrayType,
  planning_time: primitiveTypes.float32,
});
const SetGoal_ResponseRefArray = ArrayType(SetGoal_ResponseRefStruct);
const SetGoal_ResponseRefStructArray = StructType({
  data: SetGoal_ResponseRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SetGoal_ResponseWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SetGoal_ResponseRefStruct(other._refObject.toObject());
      this._wrapperFields.path_ids = Int32Wrapper.createArray();
      this._wrapperFields.path_ids.copy(other._wrapperFields.path_ids);
      this._wrapperFields.path_poses = PoseWrapper.createArray();
      this._wrapperFields.path_poses.copy(other._wrapperFields.path_poses);
      this._planning_timeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SetGoal_ResponseRefStruct();
    this._wrapperFields.path_ids = Int32Wrapper.createArray();
    this._wrapperFields.path_poses = PoseWrapper.createArray();
    this._planning_timeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SetGoal_ResponseWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SetGoal_ResponseArrayWrapper;
  }
  static get ArrayType() {
    return SetGoal_ResponseArrayWrapper;
  }
  static get refObjectArrayType() {
    return SetGoal_ResponseRefStructArray
  }
  static get refObjectType() {
    return SetGoal_ResponseRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._planning_timeIntialized) {
        throw new TypeError('Invalid argument: planning_time in SetGoal_Response');
      }
    this._wrapperFields.path_ids.freeze(own, checkConsistency);
    this._refObject.path_ids = this._wrapperFields.path_ids.refObject;
    if (own) {
      this._wrapperFields.path_ids.fill(Int32Array.from([]));
      this._wrapperFields.path_ids.freeze(own, checkConsistency);
      this._refObject.path_ids = this._wrapperFields.path_ids.refObject;
    }
    this._wrapperFields.path_poses.freeze(own, checkConsistency);
    this._refObject.path_poses = this._wrapperFields.path_poses.refObject;
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.path_ids.copyRefObject(refObject.path_ids);
    this._wrapperFields.path_poses.copyRefObject(refObject.path_poses);
    this._planning_timeIntialized = true;
    this._refObject.planning_time = refObject.planning_time;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.path_ids.size != 0) {
      Int32Wrapper.ArrayType.freeArray(refObject.path_ids);
      if (Int32Wrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.path_ids, Int32Wrapper.refObjectArrayType, 'data');
      }
    }
    if (refObject.path_poses.size != 0) {
      PoseWrapper.ArrayType.freeArray(refObject.path_poses);
      if (PoseWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.path_poses, PoseWrapper.refObjectArrayType, 'data');
      }
    }
  }
  static destoryRawROS(msg) {
    SetGoal_ResponseWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'SetGoal_Response'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get path_ids() {
    return this._wrapperFields['path_ids'].data;
  }
  set path_ids(value) {
    this._wrapperFields['path_ids'].fill(value);
  }
  get path_poses() {
    return this._wrapperFields.path_poses;
  }
  set path_poses(value) {
    this._wrapperFields.path_poses.fill(value);
  }
  get planning_time() {
    if (!this._planning_timeIntialized) {
      return undefined;
    }
    return this._refObject.planning_time;
  }
  set planning_time(value) {
    this._planning_timeIntialized = true;
    this._refObject.planning_time = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SetGoal_ResponseRefStruct(refObject.toObject());
    this._wrapperFields.path_ids.copyRefObject(this._refObject.path_ids);
    this._wrapperFields.path_poses.copyRefObject(this._refObject.path_poses);
    this._planning_timeIntialized = true;
  }
  copy(other) {
    this._refObject = new SetGoal_ResponseRefStruct(other._refObject.toObject());
    this._wrapperFields.path_ids.copy(other._wrapperFields.path_ids);
    this._wrapperFields.path_poses.copy(other._wrapperFields.path_poses);
    this._planning_timeIntialized = true;
  }
  static get classType() {
    return SetGoal_ResponseWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"path_ids","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"path_poses","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"Pose","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planning_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"SetGoal_Response","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SetGoal_Response"};
  }
  hasMember(name) {
    let memberNames = ["path_ids","path_poses","planning_time"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SetGoal_ResponseArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SetGoal_ResponseWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SetGoal_ResponseWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGoal_ResponseArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SetGoal_ResponseArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SetGoal_ResponseRefArray(size);
    this._refObject = new SetGoal_ResponseRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SetGoal_ResponseWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SetGoal_ResponseArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SetGoal_ResponseArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SetGoal_ResponseWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SetGoal_ResponseWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SetGoal_ResponseArrayWrapper;
  }
}
module.exports = SetGoal_ResponseWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "path_ids",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "path_poses",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "Pose",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planning_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "SetGoal_Response",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SetGoal_Response"
}
*/
