// rclnodejs/generated/rtabmap_msgs/rtabmap_msgs__srv__CleanupLocalGrids_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const CleanupLocalGrids_RequestRefStruct = StructType({
  radius: primitiveTypes.int32,
  filter_scans: primitiveTypes.bool,
});
const CleanupLocalGrids_RequestRefArray = ArrayType(CleanupLocalGrids_RequestRefStruct);
const CleanupLocalGrids_RequestRefStructArray = StructType({
  data: CleanupLocalGrids_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CleanupLocalGrids_RequestWrapper {
  constructor(other, willCheckConsistency = false) {
    this._wrapperFields = {};
    this._willCheckConsistency = willCheckConsistency;
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CleanupLocalGrids_RequestRefStruct(other._refObject.toObject());
      this._radiusIntialized = true;
      this._filter_scansIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CleanupLocalGrids_RequestRefStruct();
    this._radiusIntialized = false;
    this._filter_scansIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CleanupLocalGrids_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CleanupLocalGrids_RequestArrayWrapper;
  }
  static get ArrayType() {
    return CleanupLocalGrids_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return CleanupLocalGrids_RequestRefStructArray
  }
  static get refObjectType() {
    return CleanupLocalGrids_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
      if (checkConsistency && !this._radiusIntialized) {
        throw new TypeError('Invalid argument: radius in CleanupLocalGrids_Request');
      }
      if (checkConsistency && !this._filter_scansIntialized) {
        throw new TypeError('Invalid argument: filter_scans in CleanupLocalGrids_Request');
      }
  }
  serialize() {
    this.freeze(/*own=*/false, this._willCheckConsistency);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._radiusIntialized = true;
    this._refObject.radius = refObject.radius;
    this._filter_scansIntialized = true;
    this._refObject.filter_scans = refObject.filter_scans;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    CleanupLocalGrids_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rtabmap_msgs', subFolder: 'srv', interfaceName: 'CleanupLocalGrids_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get radius() {
    if (!this._radiusIntialized) {
      return undefined;
    }
    return this._refObject.radius;
  }
  set radius(value) {
    this._radiusIntialized = true;
    this._refObject.radius = value;
  }
  get filter_scans() {
    if (!this._filter_scansIntialized) {
      return undefined;
    }
    return this._refObject.filter_scans;
  }
  set filter_scans(value) {
    this._filter_scansIntialized = true;
    this._refObject.filter_scans = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CleanupLocalGrids_RequestRefStruct(refObject.toObject());
    this._radiusIntialized = true;
    this._filter_scansIntialized = true;
  }
  copy(other) {
    this._refObject = new CleanupLocalGrids_RequestRefStruct(other._refObject.toObject());
    this._radiusIntialized = true;
    this._filter_scansIntialized = true;
  }
  static get classType() {
    return CleanupLocalGrids_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"radius","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"filter_scans","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rtabmap_msgs","type":"CleanupLocalGrids_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CleanupLocalGrids_Request"};
  }
  hasMember(name) {
    let memberNames = ["radius","filter_scans"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CleanupLocalGrids_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CleanupLocalGrids_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CleanupLocalGrids_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CleanupLocalGrids_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CleanupLocalGrids_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CleanupLocalGrids_RequestRefArray(size);
    this._refObject = new CleanupLocalGrids_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CleanupLocalGrids_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CleanupLocalGrids_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CleanupLocalGrids_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CleanupLocalGrids_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CleanupLocalGrids_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CleanupLocalGrids_RequestArrayWrapper;
  }
}
module.exports = CleanupLocalGrids_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "radius",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "filter_scans",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rtabmap_msgs",
    "type": "CleanupLocalGrids_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CleanupLocalGrids_Request"
}
*/
